<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Correspondence Analysis in R • PCAworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Correspondence Analysis in R">
<meta property="og:description" content="PCAworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PCAworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a_Intro.html">A short PCA vignette - Wine Example</a>
    </li>
    <li>
      <a href="../articles/b_PCA.html">Principal Component Analysis in R</a>
    </li>
    <li>
      <a href="../articles/c_COA.html">Correspondence Analysis in R</a>
    </li>
    <li>
      <a href="../articles/d_sc_example.html">PCA example: scRNAseq</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/aedin/PCAworkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="c_COA_files/header-attrs-2.9/header-attrs.js"></script><script src="c_COA_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="c_COA_files/d3v7-7.0.0/d3.v7.min.js"></script><script src="c_COA_files/d3v7-7.0.0/d3-legend.min.js"></script><script src="c_COA_files/d3v7-7.0.0/d3-labeler.js"></script><link href="c_COA_files/d3.lasso-plugin-1.0.0/lasso.css" rel="stylesheet">
<script src="c_COA_files/d3.lasso-plugin-1.0.0/lasso.js"></script><link href="c_COA_files/scatterD3-1.0.0/scatterD3.css" rel="stylesheet">
<script src="c_COA_files/scatterD3-1.0.0/scatterD3-utils.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-scales.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-dims.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-menu.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-axes.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-dots.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-arrows.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-labels.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-labels-lines.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-lines.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-ellipses.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-zoom.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-legend.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-lasso.js"></script><script src="c_COA_files/scatterD3-1.0.0/scatterD3-exports.js"></script><script src="c_COA_files/scatterD3-binding-1.0.0/scatterD3.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Correspondence Analysis in R</h1>
                        <h4 class="author">Aedin Culhane</h4>
            
            <h4 class="date">July 15, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/aedin/PCAworkshop/blob/master/vignettes/c_COA.Rmd"><code>vignettes/c_COA.Rmd</code></a></small>
      <div class="hidden name"><code>c_COA.Rmd</code></div>

    </div>

    
    
<div id="necessary-r-libraries" class="section level3">
<h3 class="hasAnchor">
<a href="#necessary-r-libraries" class="anchor"></a>Necessary R libraries</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Install, if necessary, and load necessary libraries and set up R session</span>
<span class="co">#if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sinhrks/ggfortify">ggfortify</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">corral</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://pbil.univ-lyon1.fr/ADE-4/">ade4</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">irlba</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/willtownes/glmpca">glmpca</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://juba.github.io/scatterD3/">scatterD3</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap">ComplexHeatmap</a></span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="background" class="section level2">
<h2 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h2>
<p>Decomposition of the Pearson residuals is a excellent method for exploratory analysis of count data and is also know as correspondence analysis. Because of its speed (which is comparable to PCA) and its performance it has been adopted by many fields. Correspondence analysis was originally proposed in 1935 (Hirschfeld, 1935), and was later developed by Benzécri as part of the French duality diagram framework for multivariate statistics (Benzécri, 1973; Holmes, 2008). His trainee Michael Greenacre later popularized its use with large, sparse count data in diverse settings and disciplines, including linguistics, business and marketing research, and archaeology (Greenacre, 1984, 2010). In numerical ecology, it is commonly applied to analyzing species abundance count matrices (M. Greenacre, 2013; Legendre et al., 1998). It has been successfully applied to microarray transcriptomics data (Fellenberg et al., 2001). and a Bioconductor version was implemented in the made4 Bioconductor package by Culhane et al., (2002) and the method was also used in joint and multi-table factorization methods (Culhane et al., 2003; Meng et al., 2014). Its first application on single cell RNA seq was in analysis of metagenomic scRNAseq microbiome census data data (McMurdie &amp; Holmes, 2013). Correspondence analysis is available in the ade4 and vegan R packages and was listed in a review of single and multi-table matrix factorization approaches (Meng C, Zeleznik OA. et al., 2016).</p>
<p>Correspondence analysis (COA) is considered a dual-scaling method, because both the rows and columns are scaled prior to singular value decomposition (SVD). Lauren Hsu provided a description of COA applied to both single and a multi-table extension to joint decomposition of single cell RNA seq in her master’s thesis (Biostatistics, Harvard School of Public Health, 2020) and has implemented a fast implementation of COA that supports sparse matrix operations and scalable svd (IRBLA) in her Bioconductor package <a href="https://www.bioconductor.org/packages/devel/bioc/html/corral.html"><code>corral</code></a>.</p>
<p>Hafemeister and Satija in their characterization of the SCTransform method propose that Pearson residuals of a regularized negative binomial model (a generalized linear model with sequencing depth as a covariate) could be used to remove technical characteristics while preserving biological heterogeneity, with the residuals used directly as input for downstream analysis, in place of log-transformed counts (Hafemeister &amp; Satija, 2019) In the Townes et al., (2019), generalized principal component analysis (GLM-PCA), a generalization of PCA to exponential family likelihoods but report that both the multinomial and Dirichlet-multinomial are computationally intractable on single cell data and propose Poisson and negative binomial likelihood models, and focusing on Poisson glmPCA due to its performance.</p>
<p>Single cell RNAseq data are counts and naturally Poisson, therefore the application of Pearson residuals by both Townes et al., (2019) and Hafemeister and Satija (2019) show how correspondence analysis has been “rediscovered.” In reviews by Abdi &amp; Valentin, 2007; Greenacre, 2010, they remark how frequently COA is rediscovered but each rediscovery enforces its importance to the field</p>
<p>In this vignette we compare different approaches for applying COA in R.</p>
</div>
<div id="toy-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#toy-dataset" class="anchor"></a>Toy Dataset</h2>
<p>The function <code>simData()</code> in the glmPCA vignette, orginally provided by <a href="https://github.com/willtownes/scrna2019/issues/2">Jake Yeung</a> created a simulated dataset with 4989 rows and 150 columns, 3 biological groups (clusters) of 50 cells each and 2 batches. 5000 rows are created but some are filtered to create a matrix of 4989 rows.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">simData</span><span class="op">&lt;-</span><span class="kw">function</span><span class="op">(</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">202</span><span class="op">)</span>
  <span class="va">ngenes</span> <span class="op">&lt;-</span> <span class="fl">5000</span> <span class="co">#must be divisible by 10</span>
  <span class="va">ngenes_informative</span><span class="op">&lt;-</span><span class="va">ngenes</span><span class="op">*</span><span class="fl">.1</span>
  <span class="va">ncells</span> <span class="op">&lt;-</span> <span class="fl">50</span> <span class="co">#number of cells per cluster, must be divisible by 2</span>
  <span class="va">nclust</span><span class="op">&lt;-</span> <span class="fl">3</span>
  <span class="co"># simulate two batches with different depths</span>
  <span class="va">batch</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, each <span class="op">=</span> <span class="va">nclust</span><span class="op">*</span><span class="va">ncells</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>
  <span class="va">ncounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html">rpois</a></span><span class="op">(</span><span class="va">ncells</span><span class="op">*</span><span class="va">nclust</span>, lambda <span class="op">=</span> <span class="fl">1000</span><span class="op">*</span><span class="va">batch</span><span class="op">)</span>
  <span class="co"># generate profiles for 3 clusters</span>
  <span class="va">profiles_informative</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">replicate</a></span><span class="op">(</span><span class="va">nclust</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">ngenes_informative</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="va">profiles_const</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span>ncol<span class="op">=</span><span class="va">nclust</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">ngenes</span><span class="op">-</span><span class="va">ngenes_informative</span><span class="op">)</span><span class="op">)</span>,<span class="va">nclust</span><span class="op">)</span><span class="op">)</span>
  <span class="va">profiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">profiles_informative</span>,<span class="va">profiles_const</span><span class="op">)</span>
  <span class="co"># generate cluster labels</span>
  <span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, each <span class="op">=</span> <span class="va">ncells</span><span class="op">)</span><span class="op">)</span>
  <span class="co"># generate single-cell transcriptomes </span>
  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/Multinom.html">rmultinom</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">ncounts</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, prob <span class="op">=</span> <span class="va">profiles</span><span class="op">[</span>,<span class="va">clust</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
  <span class="op">}</span><span class="op">)</span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"cell"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
  <span class="co"># clean up rows</span>
  <span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span>
  <span class="va">sz</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>
  <span class="va">Ycpm</span><span class="op">&lt;-</span><span class="fl">1e6</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">/</span><span class="va">sz</span><span class="op">)</span>
  <span class="va">Yl2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="va">Ycpm</span><span class="op">)</span>
  <span class="va">z</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">sz</span><span class="op">)</span>
  <span class="va">pz</span><span class="op">&lt;-</span><span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="va">Y</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span>
  <span class="va">cm</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>total_counts<span class="op">=</span><span class="va">sz</span>,zero_frac<span class="op">=</span><span class="va">pz</span>,clust<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>,batch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Y<span class="op">=</span><span class="va">Y</span>, pheno<span class="op">=</span><span class="va">cm</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">simDataObj</span> <span class="op">&lt;-</span><span class="fu">simData</span><span class="op">(</span><span class="op">)</span>
<span class="va">mat</span><span class="op">=</span><span class="va">simDataObj</span><span class="op">$</span><span class="va">Y</span>
<span class="va">pheno</span><span class="op">=</span><span class="va">simDataObj</span><span class="op">$</span><span class="va">pheno</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4989  150</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4989  150</code></pre>
<p>There are 3 clusters and 2 batches</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">clust</span>, <span class="va">pheno</span><span class="op">$</span><span class="va">batch</span><span class="op">)</span></code></pre></div>
<pre><code>##    
##      1  2
##   1 27 23
##   2 25 25
##   3 23 27</code></pre>
</div>
<div id="coa-background" class="section level1">
<h1 class="hasAnchor">
<a href="#coa-background" class="anchor"></a>COA Background</h1>
<p>So what does correspondence analysis do and how is it related to decomposition of the Pearson residuals?</p>
<p>As mentioned above a critical step in decomposition is choosing the form of the data that makes more sense. SVD will detect linear vectors that can the most variance in the data. So if the variance is skewed or varies with the mean, it needs normalization.</p>
<p>I will describe it in two ways, you could consider the expected value of an element (<span class="math inline">\(X_{ij}\)</span>) in a matrix N to be the product of its row (<span class="math inline">\(r_i\)</span>) and column (<span class="math inline">\(c_j\)</span>) weight, where the row weight is the <span class="math inline">\(\frac{rowsum}{total sum}\)</span> and thus represents the contribution of that row to the total matrix. Equally, the column weight is the <span class="math inline">\(\frac{colsum}{total sum}\)</span>, and is the contribution of that column to the total. Sometimes the row and column weight are called the row and column mass respectively. The expected weight of every element is the outer product of the row and column weights</p>
<p>In COA, the data are treated like a contingency table so the residuals are the difference between the observed data and the expected, under the assumption that there is no relationship. The Pearson <span class="math inline">\(\chi^2\)</span> statistic is: <span class="math display">\[\frac{observed-expected}{\sqrt{expected}}\]</span> The <span class="math inline">\(\chi^2\)</span> transformed matrix is then decomposed with SVD.</p>
<div class="figure">
<img src="image/coa.png" alt="COA" width="100%"><p class="caption">
COA
</p>
</div>
<p>A much more detailed explanation of correspondence analysis is provided at: <a href="https://www.displayr.com/math-correspondence-analysis/" class="uri">https://www.displayr.com/math-correspondence-analysis/</a></p>
<div id="expected-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#expected-matrix" class="anchor"></a>Expected matrix</h2>
<p>The input matrix must be positive count data.</p>
<p>The expected value for each element of the matrix is the product of the row and column weight. This can be easily calculated</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rw</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="co"># row weights</span>
<span class="va">cw</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>  <span class="co"># column weights</span>
<span class="va">expectedValues</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/outer.html">outer</a></span><span class="op">(</span><span class="va">cw</span>, <span class="va">rw</span><span class="op">)</span>
<span class="co">#dim(expw)</span>
<span class="va">expectedValues</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></code></pre></div>
<pre><code>##              gene_1       gene_2       gene_3
## cell_1 1.433571e-06 6.881143e-07 4.205143e-07
## cell_2 1.446955e-06 6.945386e-07 4.244403e-07</code></pre>
<p>Objective values</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></code></pre></div>
<pre><code>##        cell_1 cell_2 cell_3
## gene_1      0      0      0
## gene_2      1      0      0
## gene_3      0      0      0</code></pre>
<p>Therefore the <span class="math inline">\(\chi^2\)</span> transformed matrix is just as stated above: <span class="math inline">\(\frac{observed-expected}{\sqrt{expected}}\)</span></p>
</div>
<div id="pca-of-toy-data-for-comparison" class="section level2">
<h2 class="hasAnchor">
<a href="#pca-of-toy-data-for-comparison" class="anchor"></a>PCA of toy data (for comparison)</h2>
<p>For comparison we will include a PCA of the 150 cases which form 3 clusters</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pca1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>, rank. <span class="op">=</span> <span class="fl">2</span>,center<span class="op">=</span><span class="cn">TRUE</span>, scale<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">pca1</span><span class="op">)</span></code></pre></div>
<pre><code>## Importance of first k=2 (out of 150) components:
##                             PC1     PC2
## Standard deviation     13.49647 8.39371
## Proportion of Variance  0.03651 0.01412
## Cumulative Proportion   0.03651 0.05063</code></pre>
<p>Plot of first two components, colored by cluster. <code>autoplot</code> is a wrapper for plotting common R objects, including output from <code>prcomp</code> and <code>princomp</code>. It requires the library <code>ggfortify</code> to run.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sinhrks/ggfortify">ggfortify</a></span><span class="op">)</span>
<span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html">autoplot</a></span><span class="op">(</span><span class="va">pca1</span>, 
                  data<span class="op">=</span><span class="va">pheno</span>, colour<span class="op">=</span><span class="st">'clust'</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Clusters"</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-7-1.png" width="1000"></p>
<p>Plot of first two components, colored by batch. PCA does not split the count data by clusters.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html">autoplot</a></span><span class="op">(</span><span class="va">pca1</span>, 
                  data<span class="op">=</span><span class="va">pheno</span>, colour<span class="op">=</span><span class="st">'batch'</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Batch"</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-8-1.png" width="1000"></p>
</div>
<div id="singular-value-decomposition-svd-irbla" class="section level2">
<h2 class="hasAnchor">
<a href="#singular-value-decomposition-svd-irbla" class="anchor"></a>Singular Value Decomposition (svd, irbla)</h2>
<p>PCA is computed by an Singular Value Decomposition (svd) of a centered or centered and scaled matrix to compute PCA of the covariance or correlation matrix. SVD can be considered a generalization of eigenvalue decomposition. In eigenvalue decomposition the D matrix is required to be square.</p>
<p>Given a matrix X of dimension n×p, SVD decomposes it to:</p>
<p><span class="math inline">\(X = U D V^{t}\)</span></p>
<p>U,V define the left and right singular values</p>
<p>U and V are square orthogonal: <span class="math inline">\(UU^{t} = I_{p}\)</span> <span class="math inline">\(VV^{t} = I_{n}\)</span></p>
<p>the paramters <code>nu</code>, <code>nv</code> define the number of left and right singular values (components). The output <code>d</code> is a vector of singular values. See the PCA tutorial for more information.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">svd1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/svd.html">svd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, nu <span class="op">=</span> <span class="fl">2</span>,nv <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">svd1</span>,<span class="va">dim</span><span class="op">)</span></code></pre></div>
<pre><code>## $d
## NULL
## 
## $u
## [1] 150   2
## 
## $v
## [1] 4989    2</code></pre>
<p>This is the exact same as a PCA, if we compare the correlation between the left singular values (150 cells)</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">svd1</span><span class="op">$</span><span class="va">u</span>, <span class="va">pca1</span><span class="op">$</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##      PC1 PC2
## [1,]   1   0
## [2,]   0   1</code></pre>
</div>
<div id="irlba" class="section level2">
<h2 class="hasAnchor">
<a href="#irlba" class="anchor"></a>irlba</h2>
<p><code>irlba</code> is a fast and memory-efficient way to compute a partial SVD and it often used in scRNAseq application.</p>
<p>When running <code>irlba</code>, you need to define the number of Lanczos iterations (iter) carried out.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">svd2</span><span class="op">&lt;-</span><span class="fu">irlba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/irlba/man/irlba.html">irlba</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, nu <span class="op">=</span> <span class="fl">2</span>,nv <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">svd2</span>,<span class="va">dim</span><span class="op">)</span></code></pre></div>
<pre><code>## $d
## NULL
## 
## $u
## [1] 150   2
## 
## $v
## [1] 4989    2
## 
## $iter
## NULL
## 
## $mprod
## NULL</code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">svd1</span><span class="op">$</span><span class="va">u</span>, <span class="va">svd2</span><span class="op">$</span><span class="va">u</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]   -1    0
## [2,]    0   -1</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">svd1</span><span class="op">$</span><span class="va">u</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">svd2</span><span class="op">$</span><span class="va">u</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">svd1</span><span class="op">$</span><span class="va">u</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">svd2</span><span class="op">$</span><span class="va">u</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-13-1.png" width="1000"></p>
</div>
<div id="coa-using-the-ade4-function--dudi-coa" class="section level2">
<h2 class="hasAnchor">
<a href="#coa-using-the-ade4-function--dudi-coa" class="anchor"></a>COA using the ade4 function. dudi.coa</h2>
<p>Correspondence analysis is implemented in <code>ade4</code> (and <code>made4</code> extends <code>ade4</code>). The implementation of Correspondence analysis in <code>ade4</code> is slower than <code>corral</code> (see below), so is not recommended for running COA on scRNAseq. However this is a small dataset, far smaller than a typical scRNAseq data.</p>
<p>The <code>ade4</code> package provides COA in the function <code>dudi.coa</code>. In the French school, dudi refers to duality diagram (Benzécri, 1973; Holmes, 2008), which is central to the a geometric school of matrix factorization.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coa1</span><span class="op">&lt;-</span><span class="fu">ade4</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ade4/man/dudi.coa.html">dudi.coa</a></span><span class="op">(</span><span class="va">mat</span>,scannf <span class="op">=</span> <span class="cn">FALSE</span>, n<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">coa1</span><span class="op">)</span></code></pre></div>
<pre><code>## Class: coa dudi
## Call: ade4::dudi.coa(df = mat, scannf = FALSE, nf = 2)
## 
## Total inertia: 3.364
## 
## Eigenvalues:
##     Ax1     Ax2     Ax3     Ax4     Ax5 
## 0.05046 0.04830 0.03039 0.03018 0.02958 
## 
## Projected inertia (%):
##     Ax1     Ax2     Ax3     Ax4     Ax5 
##  1.5003  1.4359  0.9035  0.8972  0.8794 
## 
## Cumulative projected inertia (%):
##     Ax1   Ax1:2   Ax1:3   Ax1:4   Ax1:5 
##   1.500   2.936   3.840   4.737   5.616 
## 
## (Only 5 dimensions (out of 149) are shown)</code></pre>
<p>The resulting row and column scores and coordinates are in the objects <code>li</code>, <code>co</code>, <code>l1</code>, <code>c1</code>, where l here refers to “lignes” or rows, and c is columns.</p>
<pre><code>coa1$li  #Row coordinates dims =c(4989,10)
coa1$co  #Col coordinates dims=c(150,10)
coa1$l1  #Row scores dims =c(4989,10)
coa1$c1  #Col scores dims=c(150,10)</code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>li<span class="op">=</span><span class="va">coa1</span><span class="op">$</span><span class="va">li</span>,  l1<span class="op">=</span><span class="va">coa1</span><span class="op">$</span><span class="va">l1</span>, co<span class="op">=</span><span class="va">coa1</span><span class="op">$</span><span class="va">co</span>, c1<span class="op">=</span><span class="va">coa1</span><span class="op">$</span><span class="va">c1</span><span class="op">)</span>, <span class="va">dim</span><span class="op">)</span></code></pre></div>
<pre><code>##        li   l1  co  c1
## [1,] 4989 4989 150 150
## [2,]    2    2   2   2</code></pre>
<p>Results of <code>ade4</code> matrix factorization can be interactively explored using <code>explor</code>, by default it plots COA results in the biplot, because the features (genes) and cases are on the same scale.</p>
<p>The coordinates are in the object co and it can be seen that COA separates by cell cluster. <code>explor</code> creates plots using <code><a href="https://juba.github.io/scatterD3//reference/scatterD3.html">scatterD3::scatterD3</a></code> which creates interactive scatter plot based on d3.js.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">explor</span><span class="fu">::</span><span class="fu"><a href="https://juba.github.io/explor//reference/prepare_results.html">prepare_results</a></span><span class="op">(</span><span class="va">coa1</span><span class="op">)</span>
<span class="fu">explor</span><span class="fu">::</span><span class="fu"><a href="https://juba.github.io/explor//reference/CA_var_plot.html">CA_var_plot</a></span><span class="op">(</span><span class="va">res</span>, var_hide <span class="op">=</span> <span class="st">"Row"</span>, col_var <span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">clust</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">scatterD3</span><span class="fu">::</span><span class="fu"><a href="https://juba.github.io/scatterD3//reference/scatterD3.html">scatterD3</a></span><span class="op">(</span><span class="va">coa1</span><span class="op">$</span><span class="va">co</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">coa1</span><span class="op">$</span><span class="va">co</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, col_var<span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">clust</span>, ylab <span class="op">=</span> <span class="st">"PC2"</span>, xlab<span class="op">=</span><span class="st">"PC1"</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-8c37c826a741a251091c" style="width:1000px;height:600px;" class="scatterD3 html-widget"></div>
<script type="application/json" data-for="htmlwidget-8c37c826a741a251091c">{"x":{"data":{"x":[-0.277069031697586,-0.304512255157383,-0.354525033930855,0.25458108975744,0.106101789290931,0.185946832681902,0.107145526392882,0.192439237908701,0.100603030372716,0.183300194481866,0.196783786000631,0.149827778952335,-0.361594543047998,0.20103043052329,0.123858611190859,0.176969818486937,0.0642773347481568,0.190228672175518,0.239243167129276,-0.271151143590118,0.116824218208399,-0.34962161616678,0.154578091647418,-0.300054649231332,-0.433945428050247,0.106713175933118,-0.302117148100988,0.135781973425562,0.175139341516996,-0.491577246842446,0.146489043120314,0.170160603374406,-0.279219095326992,-0.14617373773448,-0.329823415219443,0.128299331340838,0.153103920789183,-0.302103859842492,0.0829591779936733,0.0239449002087134,-0.33174818011711,0.199347880154645,0.06906253101604,-0.291918553703537,0.172483494326062,0.132756433190315,0.196883864830933,0.192062858848358,0.138796952497364,0.0722293270422336,0.187369724248566,-0.301146539479928,0.24770636981532,0.129907607110331,-0.420504231129043,-0.362580239075117,0.104886121804934,0.0556113228240483,0.155018670712823,0.249772140988269,0.217598727730445,0.137331299287868,0.20317998919531,0.0799898758230866,0.175226336934857,-0.271553136196403,-0.322482772651841,0.147541547177562,-0.268769423503742,0.174627386230336,-0.236301114628944,-0.389050984121145,-0.277221597071536,0.170942053109996,0.200614181729471,-0.361871295244257,0.108117729500084,0.14624721318423,0.167860146899237,-0.330849649627242,0.170823258865842,-0.311169977872528,0.150004622291398,-0.200766451859187,0.224451262513622,-0.269074596153582,-0.353255816918655,-0.357105896745463,-0.393863042108452,-0.246270660672435,-0.332236913264357,0.19373016895399,0.119217669363122,0.18769293343957,0.129969188042955,0.129953761735191,-0.293695373622537,0.156431621044623,0.133385642453424,0.123418688592349,-0.38351580901641,-0.227987050039684,0.222578034508403,-0.270362076133703,0.098626709653938,0.165960923587643,-0.293730173728269,0.12005519285509,0.134915446877763,0.175189421818079,0.164832356575688,0.0737522149943065,-0.284136769036724,0.155693761463572,0.199645885425849,0.178222631478283,0.158070246410773,0.196261018009132,0.142246812271212,0.195559200981878,0.149162786350299,0.139998780902364,0.151965406839317,0.17966889658285,0.0946298442577415,0.174529270168468,-0.387957408041189,0.214465777310038,0.149280745651912,0.179525397931565,-0.332140241003829,-0.340695445353939,0.0826354157978798,0.183794904776241,0.163172277292691,0.143632171739253,0.162973973808729,0.153731489482732,-0.29517063465162,0.14920952139837,0.261909788275749,0.181635407528837,-0.279723201042633,-0.2354818474094,-0.342345195089254,-0.300428062560055,0.0915373396993332,-0.196729889991756,0.124339478194959,0.157798935062499],"y":[-0.00761930517897725,-0.0417671026454214,0.0248639906204205,0.223870174119071,-0.191036573103976,-0.349166249545503,-0.228830036083895,-0.226132422109938,0.211757245745763,-0.422436565297335,-0.400461300183906,-0.261043801441728,0.0439234206575852,-0.253310989878728,0.282630403881099,0.280654611220956,-0.242435120242368,-0.276394579851912,0.309406853263505,-0.00728483074043391,-0.271805433774082,0.093030935745186,0.282568850314032,-0.011813310543045,-0.0796742474613071,-0.163536611574384,-0.0578872897923906,0.388035906153021,0.29365342840896,-0.0321198187894336,-0.311999080140891,-0.207625956959568,-0.0675858009267707,-0.041197442009584,-0.034065905541552,-0.39601147581354,0.125535475854963,0.0199303708524559,-0.209181133129207,0.225283355110551,0.00339001468498906,-0.242662930658786,-0.292726924419276,-0.0715567463399524,-0.388897911889242,-0.162140562087827,0.34424524182795,0.352910720273077,0.29631969723488,0.222795916904849,0.325700889588306,-0.0179906207805447,0.18989708822529,0.218952985204635,0.0651623452276797,0.0207814339199246,-0.199228372605705,0.196487399516392,0.147662928409747,0.359576342783003,0.313741225903791,-0.178929830494731,-0.315915365004104,-0.149777389454971,-0.291368793333487,-0.0691175869245484,0.0478432806979671,-0.238814891243254,-0.00966986601937856,0.306038233859399,0.0283095119445827,-0.0270217865413566,-0.0436735077658822,-0.297886087946894,0.372844176384521,0.038405451183392,-0.179846413638027,0.257241446095358,-0.301501258149746,-0.0382085840642468,-0.305920437159088,0.0623095241387087,-0.298142472481574,-0.0532037153947088,0.271965827735693,0.063482546124684,0.00702645893549956,0.0719099872040799,0.011936657683843,-0.00758499985867474,-0.00180144079015531,-0.211917313350226,-0.289816449183883,-0.355191722454662,0.232673148349836,0.218904154866688,-0.0125652864402916,0.333360283477857,0.209044919069926,0.30588749310698,-0.0394541483254556,0.0774539464526385,0.290582916319931,-0.00230293921712046,-0.262228810304401,0.247768104782729,0.00822395997550119,0.190466223212091,-0.324789355224111,0.286892569065357,-0.361258040618893,-0.253730345482943,0.0157572513991313,0.19300478053157,0.162399353076452,-0.271892266479738,-0.286638822100598,-0.275097122494574,-0.196494305455543,0.285818596296096,0.218167477383944,-0.200567143981487,-0.284445793617117,0.272661548465337,0.114786434775308,0.244585054526047,0.0295232478045474,-0.259880173894861,-0.295739127817085,0.30005039532902,-0.00853756308256165,0.0237300590928791,0.144179278412984,0.241474071753026,0.266730825312856,0.260579616596882,0.29288866114526,0.265636583309971,0.0550636160388345,-0.277509294629798,-0.360030030841667,0.263873905042602,0.0171659554921945,-0.0110175994056303,0.0089180370138361,-0.00844583218833976,-0.199070040116357,0.0135664857746345,-0.223550810781084,0.195805927465396],"col_var":["2","2","2","3","1","1","1","1","3","1","1","1","2","1","3","3","1","1","3","2","1","2","3","2","2","1","2","3","3","2","1","1","2","2","2","1","3","2","1","3","2","1","1","2","1","1","3","3","3","3","3","2","3","3","2","2","1","3","3","3","3","1","1","1","1","2","2","1","2","3","2","2","2","1","3","2","1","3","1","2","1","2","1","2","3","2","2","2","2","2","2","1","1","1","3","3","2","3","3","3","2","2","3","2","1","3","2","3","1","3","1","1","2","3","3","1","1","1","1","3","3","1","1","3","3","3","2","1","1","3","2","2","3","3","3","3","3","3","2","1","1","3","2","2","2","2","1","2","1","3"],"key_var":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150]},"settings":{"x_log":false,"y_log":false,"labels_size":10,"labels_positions":null,"point_size":64,"point_opacity":1,"opacities":null,"hover_size":1,"hover_opacity":null,"xlab":"PC1","ylab":"PC2","has_labels":false,"col_lab":"pheno$clust","col_continuous":false,"col_levels":["1","2","3"],"colors":null,"ellipses":false,"ellipses_data":[],"symbol_lab":"NULL","symbol_levels":null,"symbols":null,"size_range":[10,300],"size_lab":"NULL","sizes":null,"opacity_lab":"NULL","opacities.1":null,"unit_circle":false,"has_color_var":true,"has_symbol_var":false,"has_size_var":false,"has_opacity_var":false,"has_url_var":false,"has_legend":true,"has_tooltips":true,"tooltip_text":null,"tooltip_position_x":"right","tooltip_position_y":"bottom","has_custom_tooltips":false,"click_callback":null,"init_callback":null,"zoom_callback":null,"zoom_on":null,"zoom_on_level":null,"disable_wheel":false,"fixed":false,"legend_width":150,"left_margin":30,"html_id":"scatterD3-UMSYZNOC","xlim":null,"ylim":null,"x_categorical":false,"y_categorical":false,"x_levels":null,"y_levels":null,"menu":true,"lasso":false,"lasso_callback":null,"dom_id_reset_zoom":"scatterD3-reset-zoom","dom_id_svg_export":"scatterD3-svg-export","dom_id_lasso_toggle":"scatterD3-lasso-toggle","transitions":false,"axes_font_size":"100%","legend_font_size":"100%","caption":null,"lines":{"slope":[0,null],"intercept":[0,0],"stroke_dasharray":[5,5]},"hashes":[]}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">coa1</span><span class="op">$</span><span class="va">co</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">coa1</span><span class="op">$</span><span class="va">co</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, col<span class="op">=</span><span class="va">pheno</span><span class="op">$</span><span class="va">clust</span>, pch<span class="op">=</span><span class="fl">19</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-18-1.png" width="1000"></p>
</div>
</div>
<div id="glmpca" class="section level1">
<h1 class="hasAnchor">
<a href="#glmpca" class="anchor"></a>glmPCA</h1>
<p>glmPCA is iterative, so in this example I will set seed so its reproducible. I discuss about the iterative nature of glmPCA below</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>
<span class="va">coa2</span><span class="op">&lt;-</span><span class="fu">glmpca</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/glmpca/man/glmpca.html">glmpca</a></span><span class="op">(</span><span class="va">mat</span>,L<span class="op">=</span><span class="fl">2</span>,fam<span class="op">=</span><span class="st">"poi"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">coa2</span><span class="op">$</span><span class="va">factors</span><span class="op">)</span>  <span class="co"># Column scores dim=c(150,   2)</span></code></pre></div>
<pre><code>## [1] 150   2</code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">coa2</span><span class="op">$</span><span class="va">loadings</span><span class="op">)</span> <span class="co"># row scoress dim =c(4989 ,   2)</span></code></pre></div>
<pre><code>## [1] 4989    2</code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">coa2</span><span class="op">$</span><span class="va">factors</span><span class="op">$</span><span class="va">dim1</span>, <span class="va">coa2</span><span class="op">$</span><span class="va">factors</span><span class="op">$</span><span class="va">dim2</span>, col<span class="op">=</span><span class="va">pheno</span><span class="op">$</span><span class="va">clust</span>, pch<span class="op">=</span><span class="fl">19</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-20-1.png" width="1000"></p>
</div>
<div id="corral" class="section level1">
<h1 class="hasAnchor">
<a href="#corral" class="anchor"></a>corral</h1>
<p><code>corral</code> from Lauren Hsu uses irlba and sparse matrix representation, providing a fast implementation correspondence analysis in the function <code>corral</code>. <code>corral</code> accepts input formats <code>SingleCellExperiment</code> (Bioconductor), <code>SparseMatrix</code>, <code>matrix</code>, <code>data.frame</code>. See the [dimension reduction vignette]<a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/corral/inst/doc/corral_dimred.html" class="uri">https://www.bioconductor.org/packages/devel/bioc/vignettes/corral/inst/doc/corral_dimred.html</a> from the package for more details.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coa3</span><span class="op">&lt;-</span><span class="fu">corral</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/corral/man/corral.html">corral</a></span><span class="op">(</span><span class="va">mat</span>,ncomp<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="va">coa3</span></code></pre></div>
<pre><code>## corral output summary===========================================
##   Output "list" includes standard coordinates (SCu, SCv),
##   principal coordinates (PCu, PCv), &amp; SVD output (u, d, v)
## Variance explained----------------------------------------------
##                           PC1  PC2
## percent.Var.explained    0.02 0.01
## cumulative.Var.explained 0.02 0.03
## 
## Dimensions of output elements-----------------------------------
##   Singular values (d) :: 2
##   Left singular vectors &amp; coordinates (u, SCu, PCu) :: 4989 2
##   Right singular vectors &amp; coordinates (v, SCv, PCv) :: 150 2
##   See corral help for details on each output element.
##   Use plot_embedding to visualize; see docs for details.
## ================================================================</code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coa3</span><span class="op">$</span><span class="va">PCu</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span>  <span class="co"># Row coordinates  dims =c(4989,10)</span></code></pre></div>
<pre><code>##             [,1]      [,2]
## gene_1 0.6098867 0.6836028
## gene_2 0.1179244 0.2704358</code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span>  <span class="co"># Col coordinates  dims=c(150,10)</span></code></pre></div>
<pre><code>##              [,1]        [,2]
## cell_1 -0.2770690 0.007619298
## cell_2 -0.3045123 0.041767103</code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, col<span class="op">=</span><span class="va">pheno</span><span class="op">$</span><span class="va">clust</span>, pch<span class="op">=</span><span class="fl">19</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-21-1.png" width="1000"></p>
</div>
<div id="comparing-the-output-of-all-3-coa-approaches" class="section level1">
<h1 class="hasAnchor">
<a href="#comparing-the-output-of-all-3-coa-approaches" class="anchor"></a>Comparing the output of all 3 COA approaches</h1>
<p>Correlation between correspondence analysis and <code>corral</code> is identical</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">coa1</span><span class="op">$</span><span class="va">co</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">coa1</span><span class="op">$</span><span class="va">co</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] -1</code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">coa1</span><span class="op">$</span><span class="va">co</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">coa1</span><span class="op">$</span><span class="va">co</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-23-1.png" width="1000"></p>
<p>Correlation between <code>glmpca</code> and <code>corral</code>. Note the PCs are flipped, PC1 in <code>corral</code> shares r=0.99 correlation with PC2 of <code><a href="https://rdrr.io/pkg/glmpca/man/glmpca.html">glmpca(type = poi)</a></code>.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">coa2</span><span class="op">$</span><span class="va">factors</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] -0.9893969</code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">coa2</span><span class="op">$</span><span class="va">factors</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.992287</code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="op">-</span><span class="va">coa2</span><span class="op">$</span><span class="va">factors</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">coa2</span><span class="op">$</span><span class="va">factors</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">coa3</span><span class="op">$</span><span class="va">PCv</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-25-1.png" width="1000"></p>
</div>
<div id="notes" class="section level1">
<h1 class="hasAnchor">
<a href="#notes" class="anchor"></a>Notes</h1>
<p>GLM-PCA provides generalized PCA for probit and logit models but it is an iterative algorithm. Between runs, it tends to flip the order of principal components.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">iter</span><span class="op">=</span><span class="fl">20</span>
<span class="va">tt</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">iter</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">glmpca</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/glmpca/man/glmpca.html">glmpca</a></span><span class="op">(</span><span class="va">mat</span>, L<span class="op">=</span><span class="fl">2</span>, fam<span class="op">=</span><span class="st">"poi"</span><span class="op">)</span><span class="op">)</span>
<span class="va">factors1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">iter</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">tt</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">factors</span><span class="op">$</span><span class="va">dim1</span><span class="op">)</span>  <span class="co"># first factor</span>
<span class="va">factors2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">iter</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">tt</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">factors</span><span class="op">$</span><span class="va">dim2</span><span class="op">)</span>  <span class="co"># Second factor</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">factors1</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">iter</span>, sep<span class="op">=</span><span class="st">"_"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">factors2</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"PC2"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">iter</span>, sep<span class="op">=</span><span class="st">"_"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">factors1</span>, ylab<span class="op">=</span><span class="st">"factor PC1"</span>, xlab<span class="op">=</span><span class="st">"iter"</span>, las<span class="op">=</span><span class="fl">2</span>, main<span class="op">=</span><span class="st">"The First PC1 in 20 iterations"</span><span class="op">)</span>  <span class="co"># dim=c(150  , iter)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-27-1.png" width="1000"></p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">factors1</span><span class="op">)</span>, name<span class="op">=</span><span class="st">"Correlations in PC1 (20 iterations)"</span><span class="op">)</span></code></pre></div>
<p><img src="c_COA_files/figure-html/unnamed-chunk-28-1.png" width="1000"></p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] ComplexHeatmap_2.9.0 scatterD3_1.0.0      glmpca_0.2.0        
##  [4] irlba_2.3.3          Matrix_1.3-4         ade4_1.7-17         
##  [7] corral_1.3.1         ggfortify_0.4.12     ggplot2_3.3.5       
## [10] knitr_1.33          
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-7                matrixStats_0.60.0         
##  [3] fs_1.5.0                    RColorBrewer_1.1-2         
##  [5] doParallel_1.0.16           rprojroot_2.0.2            
##  [7] GenomeInfoDb_1.29.3         tools_4.1.0                
##  [9] bslib_0.2.5.1               utf8_1.2.2                 
## [11] R6_2.5.0                    DBI_1.1.1                  
## [13] BiocGenerics_0.39.1         colorspace_2.0-2           
## [15] GetoptLong_1.0.5            withr_2.4.2                
## [17] tidyselect_1.1.1            gridExtra_2.3              
## [19] compiler_4.1.0              textshaping_0.3.5          
## [21] Biobase_2.53.0              Cairo_1.5-12.2             
## [23] desc_1.3.0                  DelayedArray_0.19.1        
## [25] labeling_0.4.2              sass_0.4.0                 
## [27] scales_1.1.1                pkgdown_1.6.1              
## [29] systemfonts_1.0.2           stringr_1.4.0              
## [31] digest_0.6.27               rmarkdown_2.9              
## [33] XVector_0.33.0              dichromat_2.0-0            
## [35] pkgconfig_2.0.3             htmltools_0.5.1.1          
## [37] MatrixGenerics_1.5.1        highr_0.9                  
## [39] fastmap_1.1.0               maps_3.3.0                 
## [41] GlobalOptions_0.1.2         htmlwidgets_1.5.3          
## [43] rlang_0.4.11                ggthemes_4.2.4             
## [45] pals_1.7                    farver_2.1.0               
## [47] shape_1.4.6                 jquerylib_0.1.4            
## [49] generics_0.1.0              jsonlite_1.7.2             
## [51] dplyr_1.0.7                 RCurl_1.98-1.3             
## [53] magrittr_2.0.1              GenomeInfoDbData_1.2.6     
## [55] transport_0.12-2            Rcpp_1.0.7                 
## [57] munsell_0.5.0               S4Vectors_0.31.0           
## [59] fansi_0.5.0                 lifecycle_1.0.0            
## [61] stringi_1.7.3               yaml_2.2.1                 
## [63] MASS_7.3-54                 SummarizedExperiment_1.23.1
## [65] zlibbioc_1.39.0             parallel_4.1.0             
## [67] crayon_1.4.1                lattice_0.20-44            
## [69] circlize_0.4.13             mapproj_1.2.7              
## [71] magick_2.7.2                pillar_1.6.2               
## [73] GenomicRanges_1.45.0        rjson_0.2.20               
## [75] codetools_0.2-18            stats4_4.1.0               
## [77] glue_1.4.2                  evaluate_0.14              
## [79] data.table_1.14.0           MultiAssayExperiment_1.19.5
## [81] png_0.1-7                   vctrs_0.3.8                
## [83] foreach_1.5.1               gtable_0.3.0               
## [85] purrr_0.3.4                 tidyr_1.1.3                
## [87] clue_0.3-59                 assertthat_0.2.1           
## [89] cachem_1.0.5                xfun_0.24                  
## [91] ragg_1.1.3                  SingleCellExperiment_1.15.1
## [93] tibble_3.1.3                iterators_1.0.13           
## [95] memoise_2.0.0               IRanges_2.27.0             
## [97] ellipse_0.4.2               cluster_2.1.2              
## [99] ellipsis_0.3.2</code></pre>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>Abdi, H., &amp; Valentin, D. (2007). Multiple Correspondence Analysis. In N. Salkind (Ed.), Encyclopedia of Measurement and Statistics (pp. 652–657). Sage Publications, Inc. <a href="https://doi.org/10.4135/9781412952644.n299" class="uri">https://doi.org/10.4135/9781412952644.n299</a></p>
<p>Benzécri, J.-P. (Ed.). (1973). L’analyse des données.</p>
<p>Busold, C. H., Winter, S., Hauser, N., Bauer, A., Dippon, J., Hoheisel, J. D., &amp; Fellenberg, K. (2005). Integration of GO annotations in Correspondence Analysis: Facilitating the interpretation of microarray data. Bioinformatics, 21(10), 2424–2429. <a href="https://doi.org/10.1093/bioinformatics/bti367" class="uri">https://doi.org/10.1093/bioinformatics/bti367</a></p>
<p>Culhane, A C, Perrière, G., &amp; Higgins, D. G. (2003). Cross-platform comparison and visualisation of gene expression data using co-inertia analysis. BMC Bioinformatics, 4(1), 59. <a href="https://doi.org/10.1186/1471-2105-" class="uri">https://doi.org/10.1186/1471-2105-</a> 4-59</p>
<p>Culhane, A. C., Perriere, G., Considine, E. C., Cotter, T. G., &amp; Higgins, D. G. (2002). Between-group analysis of microarray data. Bioinformatics, 18(12), 1600–1608. <a href="https://doi.org/10.1093/bioinformatics/18.12.1600" class="uri">https://doi.org/10.1093/bioinformatics/18.12.1600</a></p>
<p>Duò, A., Robinson, M. D., &amp; Soneson, C. (2018). A systematic performance evaluation of clustering methods for single cell RNA-seq data. F1000Research, 7, 1141. <a href="https://doi.org/10.12688/f1000research.15666.2" class="uri">https://doi.org/10.12688/f1000research.15666.2</a></p>
<p>Fellenberg, K., Hauser, N. C., Brors, B., Neutzner, A., Hoheisel, J. D., &amp; Vingron, M. (2001). Correspondence analysis applied to microarray data. Proceedings of the National Academy of Sciences, 98(19), 10781–10786. <a href="https://doi.org/10.1073/pnas.181597298" class="uri">https://doi.org/10.1073/pnas.181597298</a></p>
<p>Grantham, R., Gautier, C., &amp; Gouy, M. (1980). Codon frequencies in 119 individual genes confirm corsistent choices of degenerate bases according to genome type. Nucleic Acids Research, 8(9), 1893–1912. <a href="https://doi.org/10.1093/nar/8.9.1893" class="uri">https://doi.org/10.1093/nar/8.9.1893</a></p>
<p>Greenacre, M. (2013). The contributions of rare objects in correspondence analysis. Ecology, 94(1), 241–249. <a href="https://doi.org/10.1890/11-1730.1" class="uri">https://doi.org/10.1890/11-1730.1</a></p>
<p>Greenacre, M. J. (1984). Theory and applications of correspondence analysis. Academic Press.</p>
<p>Greenacre, M. J. (2010). Correspondence analysis: Correspondence analysis. Wiley Interdisciplinary Reviews: Computational Statistics, 2(5), 613–619. <a href="https://doi.org/10.1002/wics.114" class="uri">https://doi.org/10.1002/wics.114</a></p>
<p>Greenacre, M., &amp; Hastie, T. (1987). The Geometric Interpretation of Correspondence Analysis. Journal of the American Statistical Association, 82(398), 437–447.https://doi.org/10.1080/01621459.1987.10478446</p>
<p>Hafemeister, C., &amp; Satija, R. (2019). Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression. Genome Biology, 20(1), 296. <a href="https://doi.org/10.1186/s13059-019-1874-1" class="uri">https://doi.org/10.1186/s13059-019-1874-1</a></p>
<p>Hicks, S. C., Townes, F. W., Teng, M., &amp; Irizarry, R. A. (2018). Missing data and technical variability in single-cell RNAsequencing experiments. Biostatistics, 19(4), 562–578. <a href="https://doi.org/10.1093/biostatistics/kxx053" class="uri">https://doi.org/10.1093/biostatistics/kxx053</a></p>
<p>Hirschfeld, H. O. (1935). A Connection between Correlation and Contingency. Mathematical Proceedings of the Cambridge Philosophical Society, 31(4), 520–524. <a href="https://doi.org/10.1017/S0305004100013517" class="uri">https://doi.org/10.1017/S0305004100013517</a></p>
<p>Holmes, S. (2008). Multivariate data analysis: The French way. ArXiv:0805.2879 [Stat], 219–233. <a href="https://doi.org/10.1214/193940307000000455" class="uri">https://doi.org/10.1214/193940307000000455</a></p>
<p>Hsu L, Culhane A. Impact of Data Preprocessing on Integrative Matrix Factorization of Single Cell Data. Front Oncol. 2020;10:973. <a href="doi:10.3389/fonc.2020.00973" class="uri">doi:10.3389/fonc.2020.00973</a>.</p>
<p>Hubert, L., &amp; Arabie, P. (1985). Comparing partitions. Journal of Classification, 2(1), 193–218. <a href="https://doi.org/10.1007/BF01908075" class="uri">https://doi.org/10.1007/BF01908075</a></p>
<p>Legendre, P., Legendre, L., Legendre, L., &amp; Legendre, L. (1998). Numerical ecology (2nd English ed). Elsevier.</p>
<p>Meng, C., Kuster, B., Culhane, A. C., &amp; Gholami, A. (2014). A multivariate approach to the integration of multi-omics datasets. BMC Bioinformatics, 15(1), 162. <a href="https://doi.org/10.1186/1471-2105-15-162" class="uri">https://doi.org/10.1186/1471-2105-15-162</a></p>
<p>Meng, C., Zeleznik, O. A., Thallinger, G. G., Kuster, B., Gholami, A. M., &amp; Culhane, A. C. (2016). Dimension reduction techniques for the integrative analysis of multi-omics data. Briefings in Bioinformatics, 17(4), 628–641. <a href="https://doi.org/10.1093/bib/bbv108" class="uri">https://doi.org/10.1093/bib/bbv108</a></p>
<p>Townes, F.W., Hicks, S.C., Aryee, M.J. et al. Feature selection and dimension reduction for single-cell RNA-Seq based on a multinomial model. Genome Biol 20, 295 (2019). <a href="https://doi.org/10.1186/s13059-019-1861-6" class="uri">https://doi.org/10.1186/s13059-019-1861-6</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aedin Culhane, Lauren Hsu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
