<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PCA example: scRNAseq • PCAworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PCA example: scRNAseq">
<meta property="og:description" content="PCAworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PCAworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a_Intro.html">A short PCA vignette - Wine Example</a>
    </li>
    <li>
      <a href="../articles/b_PCA.html">Principal Component Analysis in R</a>
    </li>
    <li>
      <a href="../articles/c_COA.html">Correspondence Analysis in R</a>
    </li>
    <li>
      <a href="../articles/d_sc_example.html">PCA example: scRNAseq</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/aedin/PCAworkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="d_sc_example_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PCA example: scRNAseq</h1>
                        <h4 class="author">Lauren Hsu</h4>
            
            <h4 class="date">7/16/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/aedin/PCAworkshop/blob/master/vignettes/d_sc_example.Rmd"><code>vignettes/d_sc_example.Rmd</code></a></small>
      <div class="hidden name"><code>d_sc_example.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Single-cell ’omics analysis enables high-resolution characterization of heterogeneous populations of cells by quantifying measurements in individual cells and thus provides a fuller, more nuanced picture into the complexity and heterogeneity between cells. However, the data also present new and significant challenges as compared to previous approaches, especially as single-cell data are much larger and sparser than data generated from bulk sequencing methods. Dimensionality reduction is a key step in the single-cell analysis to address the high dimensionality and sparsity of these data, and to enable the application of more complex, computationally expensive downstream pipelines.</p>
<p>In this example, we will use the <code>Zhengmix4eq</code> benchmarking dataset from the <a href="https://bioconductor.org/packages/release/data/experiment/html/DuoClustering2018.html"><code>DuoClustering2018</code></a> package.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">corral</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">uwot</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DuoClustering2018</span>)
<span class="no">zm4eq.sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DuoClustering2018/man/sce_full_Zhengmix.html">sce_full_Zhengmix4eq</a></span>()</pre></body></html></div>
<p>The data loads as a <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code></a> object:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">zm4eq.sce</span></pre></body></html></div>
<pre><code>## class: SingleCellExperiment 
## dim: 15568 3994 
## metadata(1): log.exprs.offset
## assays(3): counts logcounts normcounts
## rownames(15568): ENSG00000237683 ENSG00000228327 ... ENSG00000215700
##   ENSG00000215699
## rowData names(10): id symbol ... total_counts log10_total_counts
## colnames(3994): b.cells1147 b.cells6276 ... regulatory.t1084
##   regulatory.t9696
## colData names(14): dataset barcode ... libsize.drop feature.drop
## reducedDimNames(2): PCA TSNE
## altExpNames(0):</code></pre>
<p>We’ll use <code>assay</code> to access the count and logcount matrices. The genes are in the rows, and the cells are in the columns.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">counts_mat</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">zm4eq.sce</span>, <span class="st">'counts'</span>)
<span class="no">logcounts_mat</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">zm4eq.sce</span>, <span class="st">'logcounts'</span>)</pre></body></html></div>
<p><code>colData</code> and <code>rowData</code> contain (and can be used to access) the cell and gene metadata, respectively.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">colData</span>(<span class="no">zm4eq.sce</span>))</pre></body></html></div>
<pre><code>## DataFrame with 6 rows and 14 columns
##               dataset          barcode     phenoid total_features
##             &lt;integer&gt;      &lt;character&gt; &lt;character&gt;      &lt;integer&gt;
## b.cells1147         1 ACGGAGGACCGAAT-1     b.cells            427
## b.cells6276         1 GCAGCGTGCCGAAT-1     b.cells            340
## b.cells6144         1 GATTTAGAGTGTAC-1     b.cells            602
## b.cells6285         1 GCAGCTCTAGAACA-1     b.cells            410
## b.cells8679         1 TCAGTGGAGTTCAG-1     b.cells            411
## b.cells6455         1 GCCTCATGCGGGAA-1     b.cells            462
##             log10_total_features total_counts log10_total_counts
##                        &lt;numeric&gt;    &lt;numeric&gt;          &lt;numeric&gt;
## b.cells1147              2.63144         1096            3.04021
## b.cells6276              2.53275          857            2.93349
## b.cells6144              2.78032         1593            3.20249
## b.cells6285              2.61384          743            2.87157
## b.cells8679              2.61490         1031            3.01368
## b.cells6455              2.66558         1043            3.01870
##             pct_counts_top_50_features pct_counts_top_100_features
##                              &lt;numeric&gt;                   &lt;numeric&gt;
## b.cells1147                    48.1752                     66.7883
## b.cells6276                    51.1085                     69.0782
## b.cells6144                    46.2649                     63.0257
## b.cells6285                    42.7995                     57.8735
## b.cells8679                    51.0184                     67.3133
## b.cells6455                    43.9118                     60.4986
##             pct_counts_top_200_features pct_counts_top_500_features
##                               &lt;numeric&gt;                   &lt;numeric&gt;
## b.cells1147                     79.2883                     100.000
## b.cells6276                     83.6639                     100.000
## b.cells6144                     74.7646                      93.597
## b.cells6285                     71.7362                     100.000
## b.cells8679                     79.5344                     100.000
## b.cells6455                     74.8802                     100.000
##             is_cell_control libsize.drop feature.drop
##                   &lt;logical&gt;    &lt;logical&gt;    &lt;logical&gt;
## b.cells1147           FALSE        FALSE        FALSE
## b.cells6276           FALSE        FALSE        FALSE
## b.cells6144           FALSE        FALSE        FALSE
## b.cells6285           FALSE        FALSE        FALSE
## b.cells8679           FALSE        FALSE        FALSE
## b.cells6455           FALSE        FALSE        FALSE</code></pre>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">rowData</span>(<span class="no">zm4eq.sce</span>))</pre></body></html></div>
<pre><code>## DataFrame with 6 rows and 10 columns
##                              id        symbol is_feature_control mean_counts
##                     &lt;character&gt;   &lt;character&gt;          &lt;logical&gt;   &lt;numeric&gt;
## ENSG00000237683 ENSG00000237683    AL627309.1              FALSE     0.00500
## ENSG00000228327 ENSG00000228327 RP11-206L10.2              FALSE     0.00200
## ENSG00000237491 ENSG00000237491 RP11-206L10.9              FALSE     0.00150
## ENSG00000225880 ENSG00000225880     LINC00115              FALSE     0.00675
## ENSG00000230368 ENSG00000230368        FAM41C              FALSE     0.00025
## ENSG00000187634 ENSG00000187634        SAMD11              FALSE     0.00050
##                 log10_mean_counts rank_counts n_cells_counts pct_dropout_counts
##                         &lt;numeric&gt;   &lt;numeric&gt;      &lt;integer&gt;          &lt;numeric&gt;
## ENSG00000237683       0.002166062      6803.0             19             99.525
## ENSG00000228327       0.000867722      4668.0              8             99.800
## ENSG00000237491       0.000650954      4132.0              6             99.850
## ENSG00000225880       0.002921638      7635.5             27             99.325
## ENSG00000230368       0.000108560       871.0              1             99.975
## ENSG00000187634       0.000217093      2159.5              2             99.950
##                 total_counts log10_total_counts
##                    &lt;numeric&gt;          &lt;numeric&gt;
## ENSG00000237683           20           1.322219
## ENSG00000228327            8           0.954243
## ENSG00000237491            6           0.845098
## ENSG00000225880           27           1.447158
## ENSG00000230368            1           0.301030
## ENSG00000187634            2           0.477121</code></pre>
<p>This dataset includes approximately 4,000 pre-sorted and annotated cells of 4 types mixed by Duo et al. in approximately equal proportions <span class="citation">(Duò, Robinson, and Soneson, n.d.)</span>. The cells were sampled from a “Massively parallel digital transcriptional profiling of single cells” <span class="citation">(Zheng et al. 2017)</span>.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>) <span class="co"># the $ operator can be used to access columns in colData</span></pre></body></html></div>
<pre><code>## 
##         b.cells  cd14.monocytes naive.cytotoxic    regulatory.t 
##             999            1000             998             997</code></pre>
<p>For each of the 4,000 cells, there are counts for over 15,000 genes.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">counts_mat</span>)</pre></body></html></div>
<pre><code>## [1] 15568  3994</code></pre>
<p>When examining a dataset like this, we might be interested in understanding more about the cells – for example, how many clusters are there and what are the cell types?</p>
<p>We might also be interested in understanding the features themselves, perhaps to identify from the 15,000 genes those that are associated wth particular clusters.</p>
<p>Matrix factorization methods like PCA enable us to look at both “row” and “column” associations.</p>
</div>
<div id="pca" class="section level2">
<h2 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h2>
<p>First, we will select features with a simple, crude approach – we’ll sort the genes by variance, then select the 500 genes with highest variance. Selecting variable genes will improve the signal to noise ratio. Moreover, from a practical perspective, this will reduce the dataset size so that analysis steps will be computationally tractable. (Typically, in actual analysis pipelines, selection of highly variable genes (HVGs) is performed using a mean-variance stabilization, for example as in the <code>scran</code> package, or in the <code>Seurat</code> packages. This is because there is a relationship between the mean and the variance of the features. See the <a href="https://osca.bioconductor.org/feature-selection.html"><em>Orchestrating Single Cell Analysis</em></a> book for more discussion on feature selection approaches.)</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">gene_vars</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">logcounts_mat</span>, <span class="kw">FUN</span> <span class="kw">=</span> <span class="no">var</span>, <span class="kw">MARGIN</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="no">gene_vars</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">gene_vars</span>, <span class="kw">decreasing</span> <span class="kw">=</span> <span class="no">T</span>)

<span class="no">counts_mat</span> <span class="kw">&lt;-</span> <span class="no">counts_mat</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">gene_vars</span>[<span class="fl">1</span>:<span class="fl">500</span>]),]
<span class="no">logcounts_mat</span> <span class="kw">&lt;-</span> <span class="no">logcounts_mat</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">gene_vars</span>[<span class="fl">1</span>:<span class="fl">500</span>]),]</pre></body></html></div>
<p>We will start by using <code>prcomp</code> to perform PCA on the counts and the logcounts, which will find cell embeddings in a lower dimensional space.</p>
<p>For purposes of comparison later, we’ll also record the runtime.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">prc_time</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">counts_prc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="no">counts_mat</span>))
<span class="no">logcounts_prc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="no">logcounts_mat</span>)</pre></body></html></div>
<p>We’ll use the <code>plot_embedding</code> function from the <code>corral</code> package to plot the cell embeddings:</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">counts_prc</span>$<span class="no">rotation</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'PCA on counts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">logcounts_prc</span>$<span class="no">rotation</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'PCA on logcounts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<p>These examples illustrate the importance of performing pre-processing – when PCA is performed on raw counts, we see an arch effect. In contrast, when performed on log-transformed* counts, there is no arch effect. See our recent <a href="https://www.frontiersin.org/articles/10.3389/fonc.2020.00973/full">mini-review article</a> in <em>Frontiers in Oncology</em> for further discussion of data pre-processing and transformations <span class="citation">(Hsu and Culhane 2020)</span>.</p>
<p>* <em>Note: while the log-transformation(</em><span class="math inline">\(log(x+1)\)</span><em>)does improve the results with PCA here, recent papers examine the theoretical basis for this transformation and question whether it is appropriate for scRNAseq data, such as Townes et al. 2019 and Hafemeister &amp; Satija 2019. Nonetheless,</em> <span class="math inline">\(log(x+1)\)</span> <em>remains a popular pre-processing choice.</em></p>
<div id="interactive-example-1" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-example-1" class="anchor"></a><em>Interactive example 1</em>
</h3>
<p>On the count data, the first two PCs contain a strong arch effect. On both plots, use the <code>xpc</code> argument in the plotting function to change which PC you’re looking at. What do the 2nd and 3rd PCs of the <strong>PCA on counts</strong> plot look like?</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">counts_prc</span>$<span class="no">rotation</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="co"># CHANGE ME</span>
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'PCA on counts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">logcounts_prc</span>$<span class="no">rotation</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="co"># CHANGE ME</span>
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'PCA on logcounts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-10-2.png" width="700"></p>
</div>
<div id="interactive-example-2" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-example-2" class="anchor"></a><em>Interactive example 2</em>
</h3>
<p>Perform the equivalent PCA without using <code>prcomp</code>.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"># your code here</pre></body></html></div>
</div>
</div>
<div id="visualizing-embeddings" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-embeddings" class="anchor"></a>Visualizing embeddings</h2>
<p>While PCA enables us to find a much lower dimensional representation, even if we are interested in just using 10 PCs, that is still hard to visualize. Uniform Manifold Approximation and Projection (UMAP) is a popular and fast dimension method that can be used on embeddings to find a further reduced dimensional representation.</p>
<p>We’ll use UMAP (<code>uwot</code>::<code>umap</code>) to improve our visualization, incorporating the first 10 components of the PCA result as performed on the logcounts:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">logcounts_prc_umap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/uwot/man/umap.html">umap</a></span>(<span class="no">logcounts_prc</span>$<span class="no">rotation</span>[,<span class="fl">1</span>:<span class="fl">10</span>], <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="fl">30</span>)
<span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">logcounts_prc_umap</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'UMAP (PCA on logcounts)'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>We can see that the cells cluster clearly by cell type.</p>
<p>What else could we have changed here?</p>
<ul>
<li><em>Experiment with what happens if you use more or fewer PCs.</em></li>
<li>
<em>Change the</em> <code>n_neighbors</code> <em>argument in the</em> <code>umap</code> <em>call and see how it changes your embedding. This parameter modulates whether more local or global structure will be preserved.</em>
</li>
<li><em>What does UMAP on the counts reduction look like? What happens if you drop the first PC (which has the strong arch effect)?</em></li>
</ul>
</div>
<div id="speeding-up-pca" class="section level2">
<h2 class="hasAnchor">
<a href="#speeding-up-pca" class="anchor"></a>Speeding up PCA</h2>
<p>This dataset is small compared to many that we encounter in biology. We are often interested in analyzing far more than 4,000 cells at one time. Further, we may want to incorporate more features. However, even on this small dataset, it takes a long time to run <code>prcomp</code> on the full dataset without any feature selection.</p>
<p><strong>How can we speed this up?</strong></p>
<p>The <code>prcomp</code> function uses an <code>svd</code> call, which by default, will compute <strong>all</strong> of the components: the number of components is the smaller dimension (number of rows or columns). We can use the <code>nv</code> (or <code>nu</code>) arguments in <code>svd</code> to control how many components are returned, but all of the components will nonetheless be computed. However, we are usually not interested in keeping all the components, so we don’t need to compute them beyond PC50. <code><a href="https://rdrr.io/pkg/irlba/man/irlba.html">irlba::irlba</a></code> uses an iterative algorithm to compute a fast approximation to singular value decomposition, and provides a much speedier alternative for when we don’t need a full decomposition.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">2020</span>)
<span class="no">mat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1</span>:<span class="fl">10</span>,<span class="fl">10000000</span>, <span class="kw">replace</span> <span class="kw">=</span> <span class="no">T</span>), <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">1000</span>)
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">svd_mat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/svd.html">svd</a></span>(<span class="no">mat</span>, <span class="kw">nv</span> <span class="kw">=</span> <span class="fl">10</span>))</pre></body></html></div>
<pre><code>##    user  system elapsed 
##   7.019   0.638   3.982</code></pre>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">irl_mat</span> <span class="kw">&lt;-</span> <span class="kw pkg">irlba</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/irlba/man/irlba.html">irlba</a></span>(<span class="no">mat</span>, <span class="kw">nv</span> <span class="kw">=</span> <span class="fl">10</span>))</pre></body></html></div>
<pre><code>##    user  system elapsed 
##   2.386   0.056   1.240</code></pre>
<p>Since this demonstration matrix is small, both <code>irlba</code> and <code>svd</code> are fast. However, we can observe that <code>svd</code> takes substantially longer than <code>irlba</code>. As datasets scale in size, this has a big impact, and the difference in speed grows. (This difference also depends on what machine you are running on – with less compute power, the difference will be starker.)</p>
<p>We can verify that <code>irlba</code> is giving us a good approximation of <code>svd</code>:</p>
<p>Comparing singular values..</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">svd_mat</span>$<span class="no">d</span></pre></body></html></div>
<pre><code>##    [1] 17398.6832   376.8733   376.1027   375.2666   374.6135   373.9239
##    [7]   373.3499   372.6006   372.3420   371.8883   371.6684   371.2383
##   [13]   370.8610   370.6192   370.4905   369.8484   369.5309   369.2013
##   [19]   368.6767   367.8656   367.6722   367.2988   367.0741   366.7980
##   [25]   366.4726   366.1458   365.8487   365.6699   365.4277   364.7888
##   [31]   364.4552   364.3522   364.1629   363.7745   363.4712   363.1159
##   [37]   362.7739   362.3892   362.1921   361.9967   361.6288   361.1042
##   [43]   360.9795   360.5981   360.4645   360.1816   359.8722   359.5901
##   [49]   359.4906   359.2501   359.0434   358.9079   358.4553   358.0188
##   [55]   357.8865   357.5333   357.3593   357.1412   356.7922   356.6963
##   [61]   356.3908   356.1439   355.8493   355.7195   355.4975   355.1989
##   [67]   354.8612   354.5331   354.5009   354.3367   353.9096   353.6392
##   [73]   353.5981   353.2290   353.0501   352.8290   352.7823   352.5621
##   [79]   352.3725   352.1943   351.8154   351.6537   351.0593   350.9020
##   [85]   350.5889   350.4618   350.1040   350.0453   349.8836   349.6280
##   [91]   349.4489   349.1660   349.0383   348.8408   348.6174   348.2136
##   [97]   348.1530   347.9450   347.8424   347.6225   347.3537   347.1141
##  [103]   346.9898   346.6695   346.5685   346.3717   346.2306   346.1234
##  [109]   345.8729   345.5619   345.2990   345.0103   344.8101   344.5352
##  [115]   344.2916   344.0542   343.8312   343.5440   343.4900   343.3615
##  [121]   343.1045   342.8969   342.6577   342.3353   342.2191   342.1004
##  [127]   341.9555   341.7866   341.6537   341.5557   341.3932   341.2543
##  [133]   340.9341   340.6901   340.4851   340.2013   340.0724   339.9107
##  [139]   339.7215   339.5511   339.4380   339.0457   338.7222   338.5861
##  [145]   338.5732   338.3829   338.1967   338.0170   337.9207   337.7477
##  [151]   337.4561   337.2502   337.0963   336.9183   336.6118   336.5675
##  [157]   336.2912   336.0463   335.8032   335.7941   335.3879   335.3165
##  [163]   335.1959   334.8782   334.8305   334.6437   334.3077   334.2376
##  [169]   334.0347   333.8465   333.7567   333.5243   333.4052   333.2500
##  [175]   333.0530   332.8486   332.7384   332.5601   332.4119   332.1030
##  [181]   332.0052   331.8817   331.7991   331.3778   331.1760   331.0176
##  [187]   330.8246   330.6660   330.5885   330.5299   330.3593   330.1312
##  [193]   329.9328   329.6997   329.6129   329.4038   329.0931   328.9439
##  [199]   328.7629   328.6092   328.4111   328.2418   328.0555   327.9822
##  [205]   327.8788   327.6635   327.5630   327.3616   327.1643   327.0219
##  [211]   326.9445   326.5944   326.3193   326.1980   325.8461   325.6764
##  [217]   325.4666   325.4124   325.2336   324.9739   324.9491   324.7053
##  [223]   324.5572   324.4002   324.3072   324.1452   323.9015   323.7991
##  [229]   323.6632   323.5364   323.2734   323.1327   323.0615   322.8111
##  [235]   322.5652   322.4412   322.2548   322.1064   321.9420   321.7547
##  [241]   321.6378   321.5479   321.4664   321.1340   321.0779   320.9161
##  [247]   320.8496   320.6721   320.4707   320.1333   319.9528   319.8027
##  [253]   319.6887   319.4980   319.4543   319.2550   319.1064   318.8620
##  [259]   318.8285   318.6593   318.5025   318.2107   318.2035   317.9611
##  [265]   317.8133   317.7375   317.6548   317.4647   317.0295   316.9117
##  [271]   316.7516   316.6494   316.5792   316.3018   316.2973   316.1341
##  [277]   316.0126   315.8014   315.3392   315.2488   315.2099   314.8994
##  [283]   314.7977   314.6574   314.4250   314.2640   314.2123   314.0053
##  [289]   313.9398   313.8154   313.6442   313.3806   313.2615   313.1145
##  [295]   312.9731   312.8193   312.6054   312.5419   312.3271   312.1204
##  [301]   312.0911   311.8983   311.8628   311.6889   311.6211   311.3473
##  [307]   311.3371   311.2245   310.9972   310.6877   310.5701   310.4700
##  [313]   310.1749   310.0926   309.9771   309.7542   309.5646   309.3989
##  [319]   309.2584   309.1673   308.9720   308.8409   308.6611   308.5449
##  [325]   308.3027   308.0805   307.9036   307.7533   307.5525   307.4351
##  [331]   307.2692   307.1593   307.1148   306.9361   306.8445   306.6220
##  [337]   306.2723   306.1653   306.0899   305.9378   305.6831   305.5834
##  [343]   305.4361   305.3693   305.2943   304.8400   304.6497   304.6034
##  [349]   304.5618   304.1863   304.1204   303.9264   303.9056   303.6682
##  [355]   303.5857   303.4121   303.1975   303.1549   303.0182   302.9292
##  [361]   302.8274   302.6395   302.3975   302.3524   302.1137   302.0521
##  [367]   301.8466   301.7292   301.5392   301.4610   301.3884   301.1875
##  [373]   300.9774   300.9067   300.8521   300.5278   300.4736   300.2491
##  [379]   300.0051   299.9290   299.7402   299.6283   299.4059   299.3108
##  [385]   299.2192   299.0852   298.8991   298.6797   298.5809   298.4584
##  [391]   298.4159   298.1594   298.0062   297.8587   297.7525   297.4841
##  [397]   297.4593   297.4099   297.2024   297.1340   296.9414   296.8401
##  [403]   296.5655   296.4964   296.3668   296.2726   296.0249   295.8231
##  [409]   295.7075   295.5046   295.4115   295.2595   295.1619   295.0313
##  [415]   294.9028   294.7821   294.6216   294.5754   294.3672   294.3570
##  [421]   294.2102   294.0181   293.8072   293.6800   293.5003   293.2114
##  [427]   293.1975   292.9314   292.8338   292.7104   292.4406   292.3594
##  [433]   292.2036   292.0974   291.8848   291.7657   291.7025   291.5024
##  [439]   291.3107   291.1617   291.0955   290.9184   290.8494   290.6996
##  [445]   290.4807   290.4443   290.3515   290.2043   289.9644   289.9281
##  [451]   289.7661   289.6615   289.4093   289.2998   289.2078   289.1424
##  [457]   288.8124   288.6118   288.5470   288.3081   288.2648   288.2167
##  [463]   287.9782   287.8751   287.8224   287.5817   287.5407   287.3164
##  [469]   287.1675   287.1092   286.9447   286.6758   286.4354   286.3227
##  [475]   286.2440   286.0107   285.8042   285.6980   285.4529   285.2318
##  [481]   285.1991   284.9608   284.8692   284.7081   284.6327   284.5734
##  [487]   284.4587   284.4434   284.2884   283.9754   283.8647   283.8539
##  [493]   283.6521   283.5585   283.4544   283.2254   283.0271   282.9746
##  [499]   282.7410   282.5102   282.3707   282.2893   282.1453   281.9310
##  [505]   281.7859   281.7638   281.5384   281.3624   281.3385   281.1034
##  [511]   280.9014   280.8407   280.6938   280.6095   280.3955   280.2368
##  [517]   280.1085   279.9740   279.7934   279.7269   279.5986   279.4528
##  [523]   279.3472   279.0553   278.9390   278.7600   278.5904   278.5516
##  [529]   278.4799   278.1632   278.0209   277.8837   277.8159   277.6914
##  [535]   277.5661   277.4719   277.3373   277.1747   277.1109   276.7372
##  [541]   276.6927   276.5713   276.4015   276.2917   276.1345   276.0207
##  [547]   275.8696   275.5971   275.5147   275.3688   275.3079   275.1162
##  [553]   274.9628   274.8831   274.5998   274.3782   274.3408   274.2651
##  [559]   274.1654   274.0083   273.8729   273.6012   273.5932   273.2766
##  [565]   273.1342   273.0634   272.9296   272.8035   272.6684   272.5947
##  [571]   272.3932   272.2675   272.1822   272.1247   272.0821   271.8550
##  [577]   271.6251   271.5414   271.3723   271.1571   270.8865   270.7485
##  [583]   270.6371   270.4717   270.4295   270.3079   270.0515   269.9810
##  [589]   269.7957   269.7088   269.3643   269.3597   269.1299   269.1077
##  [595]   269.0154   268.9784   268.6100   268.5616   268.4620   268.3490
##  [601]   268.1309   268.0188   267.8900   267.7948   267.6682   267.4940
##  [607]   267.4447   267.1024   267.0204   266.7761   266.5602   266.4182
##  [613]   266.3782   266.3010   266.1232   265.9917   265.8141   265.6747
##  [619]   265.5400   265.4404   265.2915   265.1955   264.9297   264.6513
##  [625]   264.5956   264.4858   264.3759   264.2756   264.1125   263.8150
##  [631]   263.6134   263.5889   263.4131   263.2573   263.1454   263.0288
##  [637]   262.8944   262.7945   262.5378   262.4297   262.2591   262.1204
##  [643]   261.9639   261.9441   261.8575   261.7621   261.7173   261.4218
##  [649]   261.2444   261.0267   260.9718   260.6883   260.4530   260.3093
##  [655]   260.2948   260.0262   259.8615   259.7743   259.6171   259.4779
##  [661]   259.3419   259.2603   259.1246   259.0306   258.7327   258.6600
##  [667]   258.5072   258.4094   258.3810   258.2407   258.0764   257.8622
##  [673]   257.6783   257.6314   257.5062   257.2222   257.1170   256.9371
##  [679]   256.7990   256.6949   256.6847   256.5788   256.3567   256.2758
##  [685]   256.0399   255.9300   255.8496   255.5393   255.5012   255.2668
##  [691]   255.1181   254.9432   254.8045   254.5657   254.3772   254.2250
##  [697]   254.1470   254.1137   254.0755   253.8847   253.6954   253.6534
##  [703]   253.5022   253.2033   253.1161   253.0193   252.8062   252.7392
##  [709]   252.5501   252.4069   252.1932   252.1036   251.9998   251.8464
##  [715]   251.8156   251.5629   251.5234   251.3775   251.2641   250.9946
##  [721]   250.8577   250.6770   250.3304   250.2104   250.1646   249.9482
##  [727]   249.8179   249.6768   249.5732   249.5157   249.4054   249.1995
##  [733]   248.9413   248.8963   248.8150   248.5873   248.4749   248.3769
##  [739]   248.0707   247.9875   247.8343   247.6541   247.5136   247.4188
##  [745]   247.1541   247.0629   247.0059   246.8830   246.8441   246.5622
##  [751]   246.4894   246.3712   246.2125   245.9158   245.7174   245.5830
##  [757]   245.5230   245.2318   245.0232   244.8361   244.7632   244.7363
##  [763]   244.6025   244.3271   244.1990   244.0083   243.8731   243.8173
##  [769]   243.4877   243.4685   243.4128   243.0381   242.9489   242.8955
##  [775]   242.7100   242.5186   242.4007   242.1209   241.9315   241.8313
##  [781]   241.7227   241.6431   241.6112   241.3487   241.1205   241.0458
##  [787]   240.8756   240.6659   240.5317   240.3904   240.1259   240.0807
##  [793]   239.8974   239.7597   239.4563   239.4140   239.1948   239.1479
##  [799]   238.8930   238.7016   238.6557   238.5964   238.2503   238.1158
##  [805]   237.8651   237.8338   237.7541   237.5639   237.3692   237.3177
##  [811]   237.0957   236.9063   236.8778   236.6017   236.4629   236.3227
##  [817]   236.1804   235.9140   235.8787   235.6617   235.3684   235.3466
##  [823]   235.2571   235.1096   234.8475   234.8028   234.3231   234.2822
##  [829]   234.2241   233.9834   233.8762   233.7003   233.5199   233.3249
##  [835]   233.2147   233.0609   232.8906   232.7957   232.7288   232.5408
##  [841]   232.3240   232.2628   232.0442   231.9807   231.6755   231.4235
##  [847]   231.2138   231.1356   231.0160   230.8516   230.6440   230.4296
##  [853]   230.2373   230.0841   230.0012   229.7970   229.7153   229.5489
##  [859]   229.3395   229.1772   228.9054   228.7885   228.5269   228.4273
##  [865]   228.2284   228.1146   227.9659   227.7197   227.5232   227.4964
##  [871]   227.2224   227.1114   227.0126   226.7591   226.6344   226.3251
##  [877]   226.1694   226.1586   225.9646   225.6085   225.5111   225.4413
##  [883]   225.2776   225.2350   225.0086   224.8550   224.6807   224.4490
##  [889]   223.8248   223.7958   223.7465   223.4970   223.4132   223.1479
##  [895]   223.0932   222.8799   222.5317   222.4219   222.2678   222.1834
##  [901]   221.9584   221.9156   221.6040   221.3486   221.2288   221.0513
##  [907]   220.8538   220.7572   220.4996   220.4022   220.1875   220.0984
##  [913]   219.9328   219.7009   219.4662   219.3463   219.1162   218.9721
##  [919]   218.7057   218.5229   218.2508   217.8974   217.8189   217.7686
##  [925]   217.5728   217.2683   216.8510   216.7608   216.6480   216.5671
##  [931]   216.2273   216.0796   216.0261   215.5422   215.5201   214.9409
##  [937]   214.8791   214.8371   214.6905   214.5690   214.0146   213.8322
##  [943]   213.7304   213.5679   213.4921   213.0809   213.0442   212.6581
##  [949]   212.5989   212.5246   211.9997   211.7949   211.6051   211.4902
##  [955]   211.2700   211.1339   210.9456   210.6377   210.4113   210.0514
##  [961]   209.8119   209.6549   209.5047   209.2597   209.1523   208.7529
##  [967]   208.7033   208.2073   207.7489   207.6147   207.5505   207.3460
##  [973]   207.1214   206.7603   206.7040   206.0802   206.0483   205.7742
##  [979]   205.3532   205.1747   204.8324   204.3453   204.1032   203.8914
##  [985]   203.7278   202.9874   202.9590   202.6939   202.2761   201.5340
##  [991]   201.2690   200.9394   200.6273   200.5001   200.2194   199.7768
##  [997]   199.2089   198.6686   197.6007   197.3553</code></pre>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">irl_mat</span>$<span class="no">d</span></pre></body></html></div>
<pre><code>##  [1] 17398.6832   376.8733   376.1027   375.2666   374.6135   373.9239
##  [7]   373.3499   372.6006   372.3417   371.8863</code></pre>
<p>Comparing cell embeddings (right singular values; v matrix) [Note, that depending on the seed, the sign of the <code>irlba</code> output may flip sign. This generally doesn’t matter for downstream analysis, unless integrating datasets.]</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">svd_mat</span>$<span class="no">v</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##              [,1]         [,2]          [,3]          [,4]          [,5]
## [1,] -0.009778136  0.019108048 -1.939095e-02  0.0026076962 -0.0122338432
## [2,] -0.010080091  0.014660172  2.341659e-03 -0.0009068255  0.0139706781
## [3,] -0.009820140  0.004629647 -4.099114e-03  0.0175370979 -0.0028319681
## [4,] -0.009839423 -0.012709697  3.692118e-03  0.0027956347 -0.0004302772
## [5,] -0.010139030  0.003275806  4.090792e-05  0.0097249414  0.0084838761</code></pre>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">irl_mat</span>$<span class="no">v</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##              [,1]         [,2]          [,3]          [,4]          [,5]
## [1,] -0.009778136 -0.019108048 -1.939095e-02  0.0026076962  0.0122338432
## [2,] -0.010080091 -0.014660172  2.341659e-03 -0.0009068255 -0.0139706781
## [3,] -0.009820140 -0.004629647 -4.099114e-03  0.0175370979  0.0028319682
## [4,] -0.009839423  0.012709697  3.692118e-03  0.0027956347  0.0004302772
## [5,] -0.010139030 -0.003275806  4.090792e-05  0.0097249414 -0.0084838761</code></pre>
<div id="interactive-example-3" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-example-3" class="anchor"></a><em>Interactive example 3</em>
</h3>
<p>As we saw above, we can achieve the same results as a <code>prcomp</code> call by using the <code>scale</code> and <code>svd</code> function. Similarly, we can substitute <code>irlba</code> for svd.</p>
<p>Run PCA as in <code>prcomp</code> using <code>irlba</code>. Compare the runtime with the <code>prcomp</code> and <code>scale</code> + <code>svd</code> approaches above.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"># as in the above examples, you can use system.time() to check runtime
# if your command is longer than 1 line, you can encase them within {},
# then put it inside the system.time() call
# e.g., system.time({set of commands
#                    that take more than 1 line})</pre></body></html></div>
</div>
</div>
<div id="brief-introduction-to-correspondence-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#brief-introduction-to-correspondence-analysis" class="anchor"></a>Brief introduction to correspondence analysis</h2>
<p>PCA is the oldest dimension reduction method, and while it is arguably the most popular, it is only one of many possible methods; there are numerous methods related to and adapted from PCA. One such method is <strong>correspondence analysis (COA)</strong>.</p>
<p>Instead of a centering or z-score transformation as in PCA, COA applies a <span class="math inline">\(\chi^2\)</span> transformation prior to performing SVD, and is appropriate for use with count data. It is popular in other fields, such as ecology, business research, and archaeology, for finding associations between rows and columns.</p>
<p>Correspondence analysis is implemented in various R packages, including:</p>
<ul>
<li>
<code>ade4</code>::<code>dudi.coa()</code>
</li>
<li>
<code>made4</code>::<code>ord(type = 'coa')</code>
</li>
<li>
<code>vegan</code>::<code>cca()</code> <em>(not to be confused with canonical correlation analysis, also abbreviated CCA)</em>
</li>
<li>
<code>FactoMineR</code>::<code>CA()</code>
</li>
</ul>
<p>However, these implementations are not designed for very large data, and thus are unwieldy for single-cell analysis. To enable the fast (and easy) application of COA to such datasets, we developed the <code>corral</code> package (which we’ve been using to plot embeddings, up til this point!).</p>
<p><a href="https://www.bioconductor.org/packages/devel/bioc/html/corral.html"><img src="image/corral_sticker.png" style="width:10.0%"></a></p>
<p><code>corral</code>::<code>corral</code> performs correspondence analysis:</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">corral_time</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">corral_counts</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/corral/man/corral.html">corral</a></span>(<span class="no">counts_mat</span>))
<span class="no">corral_logcounts</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/corral/man/corral.html">corral</a></span>(<span class="no">logcounts_mat</span>)
<span class="no">corral_logcounts</span></pre></body></html></div>
<pre><code>## corral output summary===========================================
##   Output "list" includes standard coordinates (SCu, SCv),
##   principal coordinates (PCu, PCv), &amp; SVD output (u, d, v)
## Variance explained----------------------------------------------
##                           PC1  PC2  PC3  PC4  PC5  PC6  PC7  PC8
## percent.Var.explained    0.13 0.04 0.02 0.01 0.01 0.01 0.00 0.00
## cumulative.Var.explained 0.13 0.17 0.19 0.20 0.20 0.21 0.21 0.22
## 
## Dimensions of output elements-----------------------------------
##   Singular values (d) :: 10
##   Left singular vectors &amp; coordinates (u, SCu, PCu) :: 500 10
##   Right singular vectors &amp; coordinates (v, SCv, PCv) :: 3994 10
##   See corral help for details on each output element.
## ================================================================</code></pre>
<p>We used the <code>irlba</code> fast SVD approximation, so <code>corral</code> runs in a fraction of the time that <code>prcomp</code> takes:</p>
<p><code>prcomp</code> runtime:</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">prc_time</span></pre></body></html></div>
<pre><code>##    user  system elapsed 
##   1.202   0.144   0.709</code></pre>
<p><code>corral</code> runtime:</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="no">corral_time</span></pre></body></html></div>
<pre><code>##    user  system elapsed 
##   0.702   0.220   0.670</code></pre>
<p>Again, we can visualize the direct output from <code>corral</code>:</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">corral_counts</span>$<span class="no">v</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'corral on counts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">corral_logcounts</span>$<span class="no">v</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'corral on logcounts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<p>In contrast to the results from PCA, we can see that there is no arch effect in the first two PCs on the count data.</p>
<p>We can also compute and visualize the UMAP embeddings:</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="no">corral_counts_umap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/uwot/man/umap.html">umap</a></span>(<span class="no">corral_counts</span>$<span class="no">v</span>, <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="fl">30</span>)
<span class="no">corral_logcounts_umap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/uwot/man/umap.html">umap</a></span>(<span class="no">corral_logcounts</span>$<span class="no">v</span>, <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="fl">30</span>)

<span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">corral_counts_umap</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'corral on counts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">corral_logcounts_umap</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'corral on logcounts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="d_sc_example_files/figure-html/unnamed-chunk-21-2.png" width="700"></p>
<div id="interactive-example-4" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-example-4" class="anchor"></a><em>Interactive example 4</em>
</h3>
<p>Load another scRNAseq benchmarking dataset from the <code>DuoClustering2018</code> package and compare the results from using different dimension reduction methods.</p>
<p>We load and pull out the logcounts matrix from set up the <code>Zhengmix4uneq</code> dataset, which is similar to the one we used above except that the cells are in unbalanced groups. Instead of loading the full count matrix as we did above, we took the pre-filtered matrix (already selected the HVGs). Feel free to select a different one; find the list of options in the documentation for the <code>DuoClustering2018</code> package.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="no">zm4uneq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DuoClustering2018/man/sce_full_Zhengmix.html">sce_filteredExpr10_Zhengmix4uneq</a></span>()
<span class="no">zm4uneq_logcounts</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">zm4uneq</span>, <span class="st">'logcounts'</span>)</pre></body></html></div>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.4 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] DuoClustering2018_1.7.0     uwot_0.1.8                 
##  [3] ggplot2_3.3.2               corral_0.99.32             
##  [5] SingleCellExperiment_1.11.6 SummarizedExperiment_1.19.6
##  [7] DelayedArray_0.15.7         matrixStats_0.56.0         
##  [9] Matrix_1.2-18               Biobase_2.49.0             
## [11] GenomicRanges_1.41.5        GenomeInfoDb_1.25.8        
## [13] IRanges_2.23.10             S4Vectors_0.27.12          
## [15] BiocGenerics_0.35.4        
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                  fs_1.4.2                     
##  [3] bit64_0.9-7.1                 httr_1.4.1                   
##  [5] rprojroot_1.3-2               tools_4.0.0                  
##  [7] backports_1.1.8               R6_2.4.1                     
##  [9] irlba_2.3.3                   DBI_1.1.0                    
## [11] colorspace_1.4-1              withr_2.2.0                  
## [13] tidyselect_1.1.0              gridExtra_2.3                
## [15] curl_4.3                      bit_1.1-15.2                 
## [17] compiler_4.0.0                desc_1.2.0                   
## [19] labeling_0.3                  scales_1.1.1                 
## [21] rappdirs_0.3.1                pkgdown_1.5.1                
## [23] stringr_1.4.0                 digest_0.6.25                
## [25] rmarkdown_2.3                 XVector_0.29.3               
## [27] dichromat_2.0-0               pkgconfig_2.0.3              
## [29] htmltools_0.5.0               fastmap_1.0.1                
## [31] dbplyr_1.4.4                  maps_3.3.0                   
## [33] rlang_0.4.7                   ggthemes_4.2.0               
## [35] pals_1.6                      RSQLite_2.2.0                
## [37] FNN_1.1.3                     shiny_1.5.0                  
## [39] farver_2.0.3                  generics_0.0.2               
## [41] mclust_5.4.6                  dplyr_1.0.0                  
## [43] RCurl_1.98-1.2                magrittr_1.5                 
## [45] GenomeInfoDbData_1.2.3        transport_0.12-2             
## [47] Rcpp_1.0.5                    munsell_0.5.0                
## [49] viridis_0.5.1                 lifecycle_0.2.0              
## [51] stringi_1.4.6                 yaml_2.2.1                   
## [53] MASS_7.3-51.6                 zlibbioc_1.35.0              
## [55] plyr_1.8.6                    BiocFileCache_1.13.0         
## [57] AnnotationHub_2.21.1          grid_4.0.0                   
## [59] blob_1.2.1                    promises_1.1.1               
## [61] ExperimentHub_1.15.0          crayon_1.3.4                 
## [63] lattice_0.20-41               mapproj_1.2.7                
## [65] knitr_1.29                    pillar_1.4.6                 
## [67] reshape2_1.4.4                glue_1.4.1                   
## [69] BiocVersion_3.12.0            evaluate_0.14                
## [71] data.table_1.12.8             BiocManager_1.30.10          
## [73] MultiAssayExperiment_1.15.2   httpuv_1.5.4                 
## [75] vctrs_0.3.2                   gtable_0.3.0                 
## [77] purrr_0.3.4                   tidyr_1.1.0                  
## [79] assertthat_0.2.1              xfun_0.15                    
## [81] mime_0.9                      xtable_1.8-4                 
## [83] RSpectra_0.16-0               later_1.1.0.1                
## [85] viridisLite_0.3.0             tibble_3.0.3                 
## [87] AnnotationDbi_1.51.1          memoise_1.1.0                
## [89] interactiveDisplayBase_1.27.5 ellipsis_0.3.1</code></pre>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-zmdata">
<p>Duò, A, MD Robinson, and C Soneson. n.d. “A Systematic Performance Evaluation of Clustering Methods for Single-Cell Rna-Seq Data [Version 2; Peer Review: 2 Approved], Journal = F1000Research, Volume = 7, Year = 2018, Number = 1141, Doi = 10.12688/f1000research.15666.2.”</p>
</div>
<div id="ref-frontiers">
<p>Hsu, Lauren L., and Aedin C. Culhane. 2020. “Impact of Data Preprocessing on Integrative Matrix Factorization of Single Cell Data.” <em>Frontiers in Oncology</em> 10: 973. <a href="https://doi.org/10.3389/fonc.2020.00973">https://doi.org/10.3389/fonc.2020.00973</a>.</p>
</div>
<div id="ref-zheng">
<p>Zheng, Grace X. Y., Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, et al. 2017. “Massively Parallel Digital Transcriptional Profiling of Single Cells.” <em>Nature Communications</em> 8 (1): 14049. <a href="https://doi.org/10.1038/ncomms14049">https://doi.org/10.1038/ncomms14049</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aedin Culhane, Lauren Hsu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
