<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PCA example: scRNAseq • PCAworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PCA example: scRNAseq">
<meta property="og:description" content="PCAworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PCAworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/PCA.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/aedin/PCAworkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="sc_example_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PCA example: scRNAseq</h1>
                        <h4 class="author">Lauren Hsu</h4>
            
            <h4 class="date">7/16/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/aedin/PCAworkshop/blob/master/vignettes/sc_example.Rmd"><code>vignettes/sc_example.Rmd</code></a></small>
      <div class="hidden name"><code>sc_example.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Single-cell ’omics analysis enables high-resolution characterization of heterogeneous populations of cells by quantifying measurements in individual cells and thus provides a fuller, more nuanced picture into the complexity and heterogeneity between cells. However, the data also present new and significant challenges as compared to previous approaches, especially as single-cell data are much larger and sparser than data generated from bulk sequencing methods. Dimensionality reduction is a key step in the single-cell analysis to address the high dimensionality and sparsity of these data, and to enable the application of more complex, computationally expensive downstream pipelines.</p>
<p>In this example, we will use the <code>Zhengmix4eq</code> benchmarking dataset from the <a href="https://bioconductor.org/packages/release/data/experiment/html/DuoClustering2018.html"><code>DuoClustering2018</code></a> package.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">corral</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">uwot</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DuoClustering2018</span>)
<span class="no">zm4eq.sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DuoClustering2018/man/sce_full_Zhengmix.html">sce_full_Zhengmix4eq</a></span>()</pre></body></html></div>
<p>The data loads as a <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code></a> object:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">zm4eq.sce</span></pre></body></html></div>
<pre><code>## class: SingleCellExperiment 
## dim: 15568 3994 
## metadata(1): log.exprs.offset
## assays(3): counts logcounts normcounts
## rownames(15568): ENSG00000237683 ENSG00000228327 ... ENSG00000215700
##   ENSG00000215699
## rowData names(10): id symbol ... total_counts log10_total_counts
## colnames(3994): b.cells1147 b.cells6276 ... regulatory.t1084
##   regulatory.t9696
## colData names(14): dataset barcode ... libsize.drop feature.drop
## reducedDimNames(2): PCA TSNE
## altExpNames(0):</code></pre>
<p>We’ll use <code>assay</code> to access the count and logcount matrices. The genes are in the rows, and the cells are in the columns.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">counts_mat</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">zm4eq.sce</span>, <span class="st">'counts'</span>)
<span class="no">logcounts_mat</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">zm4eq.sce</span>, <span class="st">'logcounts'</span>)</pre></body></html></div>
<p><code>colData</code> and <code>rowData</code> contain (and can be used to access) the cell and gene metadata, respectively.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">colData</span>(<span class="no">zm4eq.sce</span>))</pre></body></html></div>
<pre><code>## DataFrame with 6 rows and 14 columns
##               dataset          barcode     phenoid total_features
##             &lt;integer&gt;      &lt;character&gt; &lt;character&gt;      &lt;integer&gt;
## b.cells1147         1 ACGGAGGACCGAAT-1     b.cells            427
## b.cells6276         1 GCAGCGTGCCGAAT-1     b.cells            340
## b.cells6144         1 GATTTAGAGTGTAC-1     b.cells            602
## b.cells6285         1 GCAGCTCTAGAACA-1     b.cells            410
## b.cells8679         1 TCAGTGGAGTTCAG-1     b.cells            411
## b.cells6455         1 GCCTCATGCGGGAA-1     b.cells            462
##             log10_total_features total_counts log10_total_counts
##                        &lt;numeric&gt;    &lt;numeric&gt;          &lt;numeric&gt;
## b.cells1147              2.63144         1096            3.04021
## b.cells6276              2.53275          857            2.93349
## b.cells6144              2.78032         1593            3.20249
## b.cells6285              2.61384          743            2.87157
## b.cells8679              2.61490         1031            3.01368
## b.cells6455              2.66558         1043            3.01870
##             pct_counts_top_50_features pct_counts_top_100_features
##                              &lt;numeric&gt;                   &lt;numeric&gt;
## b.cells1147                    48.1752                     66.7883
## b.cells6276                    51.1085                     69.0782
## b.cells6144                    46.2649                     63.0257
## b.cells6285                    42.7995                     57.8735
## b.cells8679                    51.0184                     67.3133
## b.cells6455                    43.9118                     60.4986
##             pct_counts_top_200_features pct_counts_top_500_features
##                               &lt;numeric&gt;                   &lt;numeric&gt;
## b.cells1147                     79.2883                     100.000
## b.cells6276                     83.6639                     100.000
## b.cells6144                     74.7646                      93.597
## b.cells6285                     71.7362                     100.000
## b.cells8679                     79.5344                     100.000
## b.cells6455                     74.8802                     100.000
##             is_cell_control libsize.drop feature.drop
##                   &lt;logical&gt;    &lt;logical&gt;    &lt;logical&gt;
## b.cells1147           FALSE        FALSE        FALSE
## b.cells6276           FALSE        FALSE        FALSE
## b.cells6144           FALSE        FALSE        FALSE
## b.cells6285           FALSE        FALSE        FALSE
## b.cells8679           FALSE        FALSE        FALSE
## b.cells6455           FALSE        FALSE        FALSE</code></pre>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">rowData</span>(<span class="no">zm4eq.sce</span>))</pre></body></html></div>
<pre><code>## DataFrame with 6 rows and 10 columns
##                              id        symbol is_feature_control mean_counts
##                     &lt;character&gt;   &lt;character&gt;          &lt;logical&gt;   &lt;numeric&gt;
## ENSG00000237683 ENSG00000237683    AL627309.1              FALSE     0.00500
## ENSG00000228327 ENSG00000228327 RP11-206L10.2              FALSE     0.00200
## ENSG00000237491 ENSG00000237491 RP11-206L10.9              FALSE     0.00150
## ENSG00000225880 ENSG00000225880     LINC00115              FALSE     0.00675
## ENSG00000230368 ENSG00000230368        FAM41C              FALSE     0.00025
## ENSG00000187634 ENSG00000187634        SAMD11              FALSE     0.00050
##                 log10_mean_counts rank_counts n_cells_counts pct_dropout_counts
##                         &lt;numeric&gt;   &lt;numeric&gt;      &lt;integer&gt;          &lt;numeric&gt;
## ENSG00000237683       0.002166062      6803.0             19             99.525
## ENSG00000228327       0.000867722      4668.0              8             99.800
## ENSG00000237491       0.000650954      4132.0              6             99.850
## ENSG00000225880       0.002921638      7635.5             27             99.325
## ENSG00000230368       0.000108560       871.0              1             99.975
## ENSG00000187634       0.000217093      2159.5              2             99.950
##                 total_counts log10_total_counts
##                    &lt;numeric&gt;          &lt;numeric&gt;
## ENSG00000237683           20           1.322219
## ENSG00000228327            8           0.954243
## ENSG00000237491            6           0.845098
## ENSG00000225880           27           1.447158
## ENSG00000230368            1           0.301030
## ENSG00000187634            2           0.477121</code></pre>
<p>This dataset includes approximately 4,000 pre-sorted and annotated cells of 4 types mixed by Duo et al. in approximately equal proportions <span class="citation">(Duò, Robinson, and Soneson, n.d.)</span>. The cells were sampled from a “Massively parallel digital transcriptional profiling of single cells” <span class="citation">(Zheng et al. 2017)</span>.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>) <span class="co"># the $ operator can be used to access columns in colData</span></pre></body></html></div>
<pre><code>## 
##         b.cells  cd14.monocytes naive.cytotoxic    regulatory.t 
##             999            1000             998             997</code></pre>
<p>For each of the 4,000 cells, there are counts for over 15,000 genes.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">counts_mat</span>)</pre></body></html></div>
<pre><code>## [1] 15568  3994</code></pre>
<p>When examining a dataset like this, we might be interested in understanding more about the cells – for example, how many clusters are there and what are the cell types?</p>
<p>We might also be interested in understanding the features themselves, perhaps to identify from the 15,000 genes those that are associated wth particular clusters.</p>
<p>Matrix factorization methods like PCA enable us to look at both “row” and “column” associations.</p>
</div>
<div id="pca" class="section level2">
<h2 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h2>
<p>First, we will select features with a simple, crude approach – we’ll sort the genes by variance, then select the 500 genes with highest variance. Selecting variable genes will improve the signal to noise ratio. Moreover, from a practical perspective, this will reduce the dataset size so that analysis steps will be computationally tractable. (Typically, in actual analysis pipelines, selection of highly variable genes (HVGs) uses a mean-variance stabilization, for example as in the <code>scran</code> package, or in the <code>Seurat</code> packages. This is because there is a relationship between the mean and the variance of the features. See the <a href="https://osca.bioconductor.org/feature-selection.html"><em>Orchestrating Single Cell Analysis</em></a> book for more discussion on feature selection approaches.)</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">gene_vars</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">logcounts_mat</span>, <span class="kw">FUN</span> <span class="kw">=</span> <span class="no">var</span>, <span class="kw">MARGIN</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="no">gene_vars</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">gene_vars</span>, <span class="kw">decreasing</span> <span class="kw">=</span> <span class="no">T</span>)

<span class="no">counts_mat</span> <span class="kw">&lt;-</span> <span class="no">counts_mat</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">gene_vars</span>[<span class="fl">1</span>:<span class="fl">500</span>]),]
<span class="no">logcounts_mat</span> <span class="kw">&lt;-</span> <span class="no">logcounts_mat</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">gene_vars</span>[<span class="fl">1</span>:<span class="fl">500</span>]),]</pre></body></html></div>
<p>We will start by using <code>prcomp</code> to perform PCA on the counts and the logcounts, which will find cell embeddings in a lower dimensional space.</p>
<p>For purposes of comparison later, we’ll also record the runtime.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">prc_time</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">counts_prc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="no">counts_mat</span>))
<span class="no">logcounts_prc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="no">logcounts_mat</span>)</pre></body></html></div>
<p>We’ll use the <code>plot_embedding</code> function from the <code>corral</code> package to plot the cell embeddings:</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">counts_prc</span>$<span class="no">rotation</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'PCA on counts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">logcounts_prc</span>$<span class="no">rotation</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'PCA on logcounts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<p>These examples illustrate the importance of performing pre-processing – when PCA is performed on raw counts, we see an arch effect. In contrast, when performed on log-transformed* counts, there is no arch effect. See our recent <a href="https://www.frontiersin.org/articles/10.3389/fonc.2020.00973/full">mini-review article</a> in <em>Frontiers in Oncology</em> for further discussion of data pre-processing and transformations <span class="citation">(Hsu and Culhane 2020)</span>.</p>
<p>* <em>Note: while the log-transformation does improve the results with PCA here, recent papers examine the theoretical basis for this transformation and question whether it is appropriate for scRNAseq data, such as Townes et al. 2019 and Hafemeister &amp; Satija 2019. Nonetheless, log(x+1) remains a popular pre-processing choice.</em></p>
<div id="interactive-example-1" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-example-1" class="anchor"></a><em>Interactive example 1</em>
</h3>
<p>On the count data, the first two PCs contain a strong arch effect. On both plots, use the <code>xpc</code> argument in the plotting function to change which PC you’re looking at. What do the 2nd and 3rd PCs of the <strong>PCA on counts</strong> plot look like?</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">counts_prc</span>$<span class="no">rotation</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="co"># CHANGE ME</span>
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'PCA on counts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">logcounts_prc</span>$<span class="no">rotation</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="co"># CHANGE ME</span>
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'PCA on logcounts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-10-2.png" width="700"></p>
</div>
<div id="interactive-example-2" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-example-2" class="anchor"></a><em>Interactive example 2</em>
</h3>
<p>Perform the equivalent PCA without using <code>prcomp</code>.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"># your code here</pre></body></html></div>
</div>
</div>
<div id="visualizing-embeddings" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-embeddings" class="anchor"></a>Visualizing embeddings</h2>
<p>While PCA enables us to find a much lower dimensional representation, even if we are interested in just using 10 PCs, that is still hard to visualize. Uniform Manifold Approximation and Projection (UMAP) is a popular and fast dimension method that can be used on embeddings to find a further reduced dimensional representation.</p>
<p>We’ll use UMAP (<code>uwot</code>::<code>umap</code>) to improve our visualization, incorporating the first 10 components of the PCA result as performed on the logcounts:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">logcounts_prc_umap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/uwot/man/umap.html">umap</a></span>(<span class="no">logcounts_prc</span>$<span class="no">rotation</span>[,<span class="fl">1</span>:<span class="fl">10</span>], <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="fl">30</span>)
<span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">logcounts_prc_umap</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'UMAP (PCA on logcounts)'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>We can see that the cells cluster clearly by cell type.</p>
<p>What else could we have changed here?</p>
<ul>
<li><em>Experiment with what happens if you use more or fewer PCs.</em></li>
<li>
<em>Change the</em> <code>n_neighbors</code> <em>argument in the</em> <code>umap</code> <em>call and see how it changes your embedding. This parameter modulates whether more local or global structure will be preserved.</em>
</li>
<li><em>What does UMAP on the counts reduction look like? What happens if you drop the first PC (which has the strong arch effect)?</em></li>
</ul>
</div>
<div id="speeding-up-pca" class="section level2">
<h2 class="hasAnchor">
<a href="#speeding-up-pca" class="anchor"></a>Speeding up PCA</h2>
<p>This dataset is small compared to many that we encounter in biology. We are often interested in analyzing far more than 4,000 cells at one time. Further, we may want to incorporate more features. However, even on this small dataset, it takes a long time to run <code>prcomp</code> on the full dataset without any feature selection.</p>
<p><strong>How can we speed this up?</strong></p>
<p>The <code>prcomp</code> function uses an <code>svd</code> call, which will compute <strong>all</strong> of the components (which, in this case, would be 500). However, the point of doing PCA is that we are only going to use a few, say 10 (almost certainly not more than 50), components. <code><a href="https://rdrr.io/pkg/irlba/man/irlba.html">irlba::irlba</a></code> also performs singular value decomposition, but uses a fast approximation.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">2020</span>)
<span class="no">mat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fl">1</span>:<span class="fl">10</span>,<span class="fl">100000</span>, <span class="kw">replace</span> <span class="kw">=</span> <span class="no">T</span>), <span class="kw">nrow</span> <span class="kw">=</span> <span class="fl">1000</span>)
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">svd_mat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/svd.html">svd</a></span>(<span class="no">mat</span>))</pre></body></html></div>
<pre><code>##    user  system elapsed 
##   0.006   0.008   0.008</code></pre>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">irl_mat</span> <span class="kw">&lt;-</span> <span class="kw pkg">irlba</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/irlba/man/irlba.html">irlba</a></span>(<span class="no">mat</span>, <span class="kw">nv</span> <span class="kw">=</span> <span class="fl">10</span>))</pre></body></html></div>
<pre><code>##    user  system elapsed 
##   0.011   0.000   0.007</code></pre>
<p>Since this demonstration matrix is small, both <code>irlba</code> and <code>svd</code> are fast. However, we can observe that <code>svd</code> takes nearly twice as long as <code>irlba</code>. As datasets scale in size, this has a big impact, and the difference in speed grows.</p>
<p>We can verify that <code>irlba</code> is giving us a good approximation of <code>svd</code>:</p>
<p>Comparing singular values..</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">svd_mat</span>$<span class="no">d</span>[<span class="fl">1</span>:<span class="fl">10</span>]</pre></body></html></div>
<pre><code>##  [1] 1742.8276  119.4202  116.7462  115.4401  114.7121  113.8254  112.2604
##  [8]  112.1026  111.1874  110.3119</code></pre>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">irl_mat</span>$<span class="no">d</span></pre></body></html></div>
<pre><code>##  [1] 1742.8276  119.4202  116.7462  115.4401  114.7121  113.8254  112.2604
##  [8]  112.1026  111.1874  110.3119</code></pre>
<p>Comparing cell embeddings (right singular values; v matrix) [Note, that depending on the seed, the sign of the <code>irlba</code> output may flip sign. This generally doesn’t matter for downstream analysis, unless integrating datasets.]</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">svd_mat</span>$<span class="no">v</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##             [,1]        [,2]        [,3]        [,4]         [,5]
## [1,] -0.09786578  0.09229720 -0.09632648  0.03355282 -0.079308781
## [2,] -0.10076497 -0.06868734  0.11303952 -0.09387799  0.026371499
## [3,] -0.09814926  0.15830518 -0.08580539  0.14594189 -0.001924932
## [4,] -0.09841613  0.02678930  0.07124697 -0.15238420 -0.131791820
## [5,] -0.10129539 -0.03626573 -0.09126326 -0.16363774  0.129290112</code></pre>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">irl_mat</span>$<span class="no">v</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##             [,1]        [,2]        [,3]        [,4]         [,5]
## [1,] -0.09786578  0.09229720  0.09632648 -0.03355282  0.079308781
## [2,] -0.10076497 -0.06868734 -0.11303952  0.09387799 -0.026371499
## [3,] -0.09814926  0.15830518  0.08580539 -0.14594189  0.001924932
## [4,] -0.09841613  0.02678930 -0.07124697  0.15238420  0.131791820
## [5,] -0.10129539 -0.03626573  0.09126326  0.16363774 -0.129290112</code></pre>
<div id="interactive-example-3" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-example-3" class="anchor"></a><em>Interactive example 3</em>
</h3>
<p>As we saw above, we can achieve the same results as a <code>prcomp</code> call by using the <code>scale</code> and <code>svd</code> function. Similarly, we can substitute <code>irlba</code> for svd.</p>
<p>Run PCA as in <code>prcomp</code> using <code>irlba</code>. Compare the runtime with the <code>prcomp</code> and <code>scale</code> + <code>svd</code> approaches above.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"># as in the above examples, you can use system.time() to check runtime
# if your command is longer than 1 line, you can encase them within {},
# then put it inside the system.time() call
# e.g., system.time({set of commands
#                    that take more than 1 line})</pre></body></html></div>
</div>
</div>
<div id="brief-introduction-to-correspondence-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#brief-introduction-to-correspondence-analysis" class="anchor"></a>Brief introduction to correspondence analysis</h2>
<p>PCA is the oldest dimension reduction method, and while it is arguably the most popular, it is only one of many possible methods; there are numerous methods related to and adapted from PCA. One such method is <strong>correspondence analysis (COA)</strong>.</p>
<p>Instead of a centering or z-score transformation as in PCA, COA applies a <span class="math inline">\(\chi^2\)</span> transformation prior to performing SVD, and is appropriate for use with count data. It is popular in other fields, such as ecology, business research, and archaeology, for finding associations between rows and columns.</p>
<p>Correspondence analysis is implemented in various R packages, including:</p>
<ul>
<li>
<code>ade4</code>::<code>dudi.coa()</code>
</li>
<li>
<code>made4</code>::<code>ord(type = 'coa')</code>
</li>
<li>
<code>vegan</code>::<code>cca()</code> <em>(not to be confused with canonical correlation analysis, also abbreviated CCA)</em>
</li>
<li>
<code>FactoMineR</code>::<code>CA()</code>
</li>
</ul>
<p>However, these implementations are not designed for very large data, and thus are unwieldy for single-cell analysis. To enable the fast (and easy) application of COA to such datasets, we developed the <code>corral</code> package (which we’ve been using to plot embeddings, up til this point!).</p>
<p><a href="https://www.bioconductor.org/packages/devel/bioc/html/corral.html"><img src="image/corral_sticker.png" style="width:10.0%"></a></p>
<p><code>corral</code>::<code>corral</code> performs correspondence analysis:</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">corral_time</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(<span class="no">corral_counts</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/corral/man/corral.html">corral</a></span>(<span class="no">counts_mat</span>))
<span class="no">corral_logcounts</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/corral/man/corral.html">corral</a></span>(<span class="no">logcounts_mat</span>)
<span class="no">corral_logcounts</span></pre></body></html></div>
<pre><code>## corral output summary===========================================
##   Output "list" includes standard coordinates (SCu, SCv),
##   principal coordinates (PCu, PCv), &amp; SVD output (u, d, v)
## Variance explained----------------------------------------------
##                           PC1  PC2  PC3  PC4  PC5  PC6  PC7  PC8
## percent.Var.explained    0.13 0.04 0.02 0.01 0.01 0.01 0.00 0.00
## cumulative.Var.explained 0.13 0.17 0.19 0.20 0.20 0.21 0.21 0.22
## 
## Dimensions of output elements-----------------------------------
##   Singular values (d) :: 10
##   Left singular vectors &amp; coordinates (u, SCu, PCu) :: 500 10
##   Right singular vectors &amp; coordinates (v, SCv, PCv) :: 3994 10
##   See corral help for details on each output element.
## ================================================================</code></pre>
<p>We used the <code>irlba</code> fast SVD approximation, so <code>corral</code> runs in a fraction of the time that <code>prcomp</code> takes:</p>
<p><code>prcomp</code> runtime:</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">prc_time</span></pre></body></html></div>
<pre><code>##    user  system elapsed 
##   1.016   0.136   0.608</code></pre>
<p><code>corral</code> runtime:</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="no">corral_time</span></pre></body></html></div>
<pre><code>##    user  system elapsed 
##   0.608   0.168   0.542</code></pre>
<p>Again, we can visualize the direct output from <code>corral</code>:</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">corral_counts</span>$<span class="no">v</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'corral on counts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">corral_logcounts</span>$<span class="no">v</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'corral on logcounts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<p>In contrast to the results from PCA, we can see that there is no arch effect in the first two PCs on the count data.</p>
<p>We can also compute and visualize the UMAP embeddings:</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="no">corral_counts_umap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/uwot/man/umap.html">umap</a></span>(<span class="no">corral_counts</span>$<span class="no">v</span>, <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="fl">30</span>)
<span class="no">corral_logcounts_umap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/uwot/man/umap.html">umap</a></span>(<span class="no">corral_logcounts</span>$<span class="no">v</span>, <span class="kw">n_neighbors</span> <span class="kw">=</span> <span class="fl">30</span>)

<span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">corral_counts_umap</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'corral on counts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/corral/man/plot_embedding.html">plot_embedding</a></span>(<span class="no">corral_logcounts_umap</span>,
               <span class="kw">xpc</span> <span class="kw">=</span> <span class="fl">1</span>,
               <span class="kw">plot_title</span> <span class="kw">=</span> <span class="st">'corral on logcounts'</span>,
               <span class="kw">color_vec</span> <span class="kw">=</span> <span class="no">zm4eq.sce</span>$<span class="no">phenoid</span>,
               <span class="kw">color_title</span> <span class="kw">=</span> <span class="st">'Cell type'</span>,
               <span class="kw">saveplot</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<p><img src="sc_example_files/figure-html/unnamed-chunk-21-2.png" width="700"></p>
<div id="interactive-example-4" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-example-4" class="anchor"></a><em>Interactive example 4</em>
</h3>
<p>Load another scRNAseq benchmarking dataset from the <code>DuoClustering2018</code> package and compare the results from using different dimension reduction methods.</p>
<p>We load and pull out the logcounts matrix from set up the <code>Zhengmix4uneq</code> dataset, which is similar to the one we used above except that the cells are in unbalanced groups. Instead of loading the full count matrix as we did above, we took the pre-filtered matrix (already selected the HVGs). Feel free to select a different one; find the list of options in the documentation for the <code>DuoClustering2018</code> package.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="no">zm4uneq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DuoClustering2018/man/sce_full_Zhengmix.html">sce_filteredExpr10_Zhengmix4uneq</a></span>()
<span class="no">zm4uneq_logcounts</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">zm4uneq</span>, <span class="st">'logcounts'</span>)</pre></body></html></div>
<div id="refs" class="references">
<div id="ref-zmdata">
<p>Duò, A, MD Robinson, and C Soneson. n.d. “A Systematic Performance Evaluation of Clustering Methods for Single-Cell Rna-Seq Data [Version 2; Peer Review: 2 Approved], Journal = F1000Research, Volume = 7, Year = 2018, Number = 1141, Doi = 10.12688/f1000research.15666.2.”</p>
</div>
<div id="ref-frontiers">
<p>Hsu, Lauren L., and Aedin C. Culhane. 2020. “Impact of Data Preprocessing on Integrative Matrix Factorization of Single Cell Data.” <em>Frontiers in Oncology</em> 10: 973. <a href="https://doi.org/10.3389/fonc.2020.00973">https://doi.org/10.3389/fonc.2020.00973</a>.</p>
</div>
<div id="ref-zheng">
<p>Zheng, Grace X. Y., Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, et al. 2017. “Massively Parallel Digital Transcriptional Profiling of Single Cells.” <em>Nature Communications</em> 8 (1): 14049. <a href="https://doi.org/10.1038/ncomms14049">https://doi.org/10.1038/ncomms14049</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aedin Culhane, Lauren Hsu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
